

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>ortho2align package &mdash; ortho2align 0.7 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> ortho2align
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <!-- Local TOC -->
              <div class="local-toc"><ul>
<li><a class="reference internal" href="#">ortho2align package</a><ul>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-ortho2align.alignment">ortho2align.alignment module</a></li>
<li><a class="reference internal" href="#module-ortho2align.cli_scripts">ortho2align.cli_scripts module</a></li>
<li><a class="reference internal" href="#module-ortho2align.fitting">ortho2align.fitting module</a></li>
<li><a class="reference internal" href="#module-ortho2align.genomicranges">ortho2align.genomicranges module</a></li>
<li><a class="reference internal" href="#module-ortho2align.orthodb">ortho2align.orthodb module</a></li>
<li><a class="reference internal" href="#module-ortho2align.parallel">ortho2align.parallel module</a></li>
<li><a class="reference internal" href="#module-ortho2align">Module contents</a></li>
</ul>
</li>
</ul>
</div>
            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">ortho2align</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>ortho2align package</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/ortho2align.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="ortho2align-package">
<h1>ortho2align package<a class="headerlink" href="#ortho2align-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-ortho2align.alignment">
<span id="ortho2align-alignment-module"></span><h2>ortho2align.alignment module<a class="headerlink" href="#module-ortho2align.alignment" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="ortho2align.alignment.Alignment">
<em class="property">class </em><code class="sig-prename descclassname">ortho2align.alignment.</code><code class="sig-name descname">Alignment</code><span class="sig-paren">(</span><em class="sig-param">HSPs</em>, <em class="sig-param">qlen=None</em>, <em class="sig-param">slen=None</em>, <em class="sig-param">filtered_HSPs=None</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.Alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>BLAST alignment of two sequences.</p>
<blockquote>
<div><p>Represents a BLAST alignment of two sequences</p>
</div></blockquote>
<dl>
<dt>(NOT an alignment of a sequence against a database!).</dt><dd><p>Contains a list of HSPs and other additional information.
Allows to read BLAST alignment from file produced with
<cite>-outfmt 7</cite>, plot an alignment map, read from and write to
dictionary representation (hence, JSON) and find best
transcripts based on some strategies.</p>
<dl>
<dt>Attributes:</dt><dd><dl class="simple">
<dt>_all_HSPs (list): list of all HSPs related</dt><dd><p>to the alignment.</p>
</dd>
<dt>HSPs (list): list of HSP instances that were</dt><dd><p>filtered by some criterion.</p>
</dd>
</dl>
<p>qlen (int): length of query sequence.
slen (int): length of subject sequence.
replace_dict (dict): field names to replace</p>
<blockquote>
<div><p>for convenience.</p>
</div></blockquote>
</dd>
</dl>
</dd>
</dl>
<dl class="method">
<dt id="ortho2align.alignment.Alignment.__eq__">
<code class="sig-name descname">__eq__</code><span class="sig-paren">(</span><em class="sig-param">other</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.Alignment.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Equality magic method.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.Alignment.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">HSPs</em>, <em class="sig-param">qlen=None</em>, <em class="sig-param">slen=None</em>, <em class="sig-param">filtered_HSPs=None</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.Alignment.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Inits Alignment class.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul>
<li><p><strong>HSPs</strong> (<em>list</em>) – list of HSP instances;</p></li>
<li><p><strong>qlen</strong> (<em>int</em>) – query sequence length
(default: None);</p></li>
<li><p><strong>slen</strong> (<em>int</em>) – subject sequence length
(default: None);</p></li>
<li><p><strong>qchrom</strong> (<em>str, int</em>) – query chromosome name (default: None).</p></li>
<li><p><strong>schrom</strong> (<em>str, int</em>) – subject chromosome name (default: None).</p></li>
<li><p><strong>qname</strong> (<em>str</em>) – query sequence name (default: None).</p></li>
<li><p><strong>sname</strong> (<em>str</em>) – subject sequence name (default: None).</p></li>
<li><p><strong>qstrand</strong> (<em>str</em>) – query sequence name.
One of “+”, “-“, “.” (default: “.”).</p></li>
<li><p><strong>sstrand</strong> (<em>str</em>) – subject sequence name.
One of “+”, “-“, “.” (default: “.”).</p></li>
<li><p><strong>filtered_HSPs</strong> (<em>list</em>) –</p>
<dl class="simple">
<dt>list of HSP instances</dt><dd><p>that were filtered by some criterion</p>
</dd>
</dl>
<p>(default: None).</p>
</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.Alignment._build_hsp_graph">
<code class="sig-name descname">_build_hsp_graph</code><span class="sig-paren">(</span><em class="sig-param">vertices</em>, <em class="sig-param">orientation</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.Alignment._build_hsp_graph" title="Permalink to this definition">¶</a></dt>
<dd><p>Builds graph from HSP list.</p>
<blockquote>
<div><p>Taken sorted list of HSPVertex instances
adds succeeding HSPs in <cite>next_vertices</cite> attribute so
as these HSPs are the closest to the given</p>
</div></blockquote>
<dl>
<dt>(only one if it does not overlap any other HSP,</dt><dd><p>more than one otherwise). Check the code and
<cite>HSP.precede</cite> for complete definition.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>vertices (list): a list of HSPVertex instances</dt><dd><p>of one direction sorted by qstart.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>a new list of HSPVertex instances.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.Alignment._find_all_transcripts">
<code class="sig-name descname">_find_all_transcripts</code><span class="sig-paren">(</span><em class="sig-param">vertices</em>, <em class="sig-param">orientation</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.Alignment._find_all_transcripts" title="Permalink to this definition">¶</a></dt>
<dd><p>Build all possible transcripts from given list
of HSPs so that HSPs follow each other on both
query and subject sequences and do not overlap
with each other.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>vertices</strong> (<em>list</em>) – a list of HSP instances
sorted by query start and query end.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of all possible transcripts.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.Alignment._find_best_transcript">
<code class="sig-name descname">_find_best_transcript</code><span class="sig-paren">(</span><em class="sig-param">vertices</em>, <em class="sig-param">orientation</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.Alignment._find_best_transcript" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds best transcript in given HSP list.</p>
<p>Utilizes dynamic programming algorithm of finding
the shortest path between two graph vertices to
get a sequence of HSPs with lowest total distance score.
Distance score is defined by <cite>HSP.distance</cite> method when
building the graph with <cite>_build_hsp_graph</cite> method.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul>
<li><p><strong>vertices</strong> (<em>list</em>) –</p>
<dl class="simple">
<dt>list of HSPVertex instances</dt><dd><p>produced by <cite>_split_orientations</cite> method</p>
</dd>
<dt>(i.e. of one orientation and sorted by qstart</dt><dd><p>and sstart);</p>
</dd>
</dl>
</li>
<li><p><strong>orientation</strong> (<em>str</em>) – one of ‘direct’, ‘reverse’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A Transcript instance with the best score.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.Alignment._split_orientations">
<code class="sig-name descname">_split_orientations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.Alignment._split_orientations" title="Permalink to this definition">¶</a></dt>
<dd><p>Splits HSPs in two groups by orientation.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.Alignment.best_transcript">
<code class="sig-name descname">best_transcript</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.Alignment.best_transcript" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds the best transcript based on its score.</p>
<p>First, finds best transcripts in direct and
reverse orientations and then choses the one
with the biggest score.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>transcript of the biggest score.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="#ortho2align.alignment.Transcript" title="ortho2align.alignment.Transcript">Transcript</a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.Alignment.cut_coordinates">
<code class="sig-name descname">cut_coordinates</code><span class="sig-paren">(</span><em class="sig-param">qleft=None</em>, <em class="sig-param">qright=None</em>, <em class="sig-param">sleft=None</em>, <em class="sig-param">sright=None</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.Alignment.cut_coordinates" title="Permalink to this definition">¶</a></dt>
<dd><p>Leaves only HSPs in certain range of coordinates.</p>
<p>Given query and subject sequence boundaries only
those HSPs are kept that lay within those boundaries.
If an HSP intersects a boundary, it is cut and its
score is reduced proportionally to the share of the cut.
If a boundary is left None than it is equal to +Inf
or -Inf for left and right boundaries, respectively.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>qleft</strong> (<em>int, float</em>) – query sequence left boundary
(default: None);</p></li>
<li><p><strong>qright</strong> (<em>int, float</em>) – query sequence right boundary
(default: None);</p></li>
<li><p><strong>sleft</strong> (<em>int, float</em>) – subject sequence left boundary
(default: None);</p></li>
<li><p><strong>sright</strong> (<em>int, float</em>) – subjecct sequence right boundary
(default: None);</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>New Alignment instance with HSPs within given
coordinates.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.Alignment.filter_by_function_score">
<code class="sig-name descname">filter_by_function_score</code><span class="sig-paren">(</span><em class="sig-param">function</em>, <em class="sig-param">score</em>, <em class="sig-param">side='g'</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.Alignment.filter_by_function_score" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter HSPs by function value of score.</p>
<p>All HSPs being stored in <cite>self.HSPs</cite>,
the method puts those HSPs in <cite>self.HSPs</cite>
that have function value of score compared to <cite>score</cite> in a
manner specified by <cite>side</cite> argument. As default (side=’g’),
function(HSP.score) have to be greater than <cite>score</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>self</strong> – Alignment instance;</p></li>
<li><p><strong>function</strong> (<em>callable</em>) – a function to apply to
HSP.score.</p></li>
<li><p><strong>score</strong> (<em>int</em>) – score to filter by.</p></li>
<li><p><strong>side</strong> (<em>str</em>) – a side to compare score and function results
by. One of:
‘g’ for greater (HSP.score &gt; score)
‘l’ for less (HSP.score &lt; score)
‘eq’ for equal (HSP.score == score)
‘neq’ for not equal (HSP.score != score)
‘geq’ for greater or equal (HSP.score &gt;= score)
‘leq’ for less or equal (HSP.score &lt;= score)
Default: ‘g’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.Alignment.filter_by_score">
<code class="sig-name descname">filter_by_score</code><span class="sig-paren">(</span><em class="sig-param">score</em>, <em class="sig-param">side='g'</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.Alignment.filter_by_score" title="Permalink to this definition">¶</a></dt>
<dd><p>Filters HSPs by score.</p>
<p>All HSPs being stored in <cite>self.HSPs</cite>,
the method puts those HSPs in <cite>self.HSPs</cite>
that have score compared to <cite>score</cite> in a
manner specified by <cite>side</cite> argument. As default (side=’g’),
HSP.score have to be greater than <cite>score</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>self</strong> – Alignment instance;</p></li>
<li><p><strong>score</strong> (<em>int</em>) – score to filter by.</p></li>
<li><p><strong>side</strong> (<em>str</em>) – a side to compare score and HSP
scores by. One of:
‘g’ for greater (HSP.score &gt; score)
‘l’ for less (HSP.score &lt; score)
‘eq’ for equal (HSP.score == score)
‘neq’ for not equal (HSP.score != score)
‘geq’ for greater or equal (HSP.score &gt;= score)
‘leq’ for less or equal (HSP.score &lt;= score)
Default: ‘g’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.Alignment.from_dict">
<em class="property">classmethod </em><code class="sig-name descname">from_dict</code><span class="sig-paren">(</span><em class="sig-param">dict_</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.Alignment.from_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Restores object from dict representation.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.Alignment.from_file_blast">
<em class="property">classmethod </em><code class="sig-name descname">from_file_blast</code><span class="sig-paren">(</span><em class="sig-param">file_object</em>, <em class="sig-param">start_type=1</em>, <em class="sig-param">end_type='inclusive'</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.Alignment.from_file_blast" title="Permalink to this definition">¶</a></dt>
<dd><p>File parser.</p>
<p>Takes file object that corresponds to
a BLAST alignment of two sequences
produced with <cite>-outfmt 7</cite> key and
parses it into an Alignment object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_object</strong> – file object.</p></li>
<li><p><strong>qchrom</strong> (<em>str, int</em>) – query chromosome name (default: None).</p></li>
<li><p><strong>schrom</strong> (<em>str, int</em>) – subject chromosome name (default: None).</p></li>
<li><p><strong>qname</strong> (<em>str</em>) – query sequence name (default: None).</p></li>
<li><p><strong>sname</strong> (<em>str</em>) – subject sequence name (default: None).</p></li>
<li><p><strong>qstrand</strong> (<em>str</em>) – query sequence name.
One of “+”, “-“, “.” (default: “.”).</p></li>
<li><p><strong>sstrand</strong> (<em>str</em>) – subject sequence name.
One of “+”, “-“, “.” (default: “.”).</p></li>
<li><p><strong>start_type</strong> (<em>int</em>) – chromosome start coordinate.
One of 0, 1 (default: 1).</p></li>
<li><p><strong>end_type</strong> (<em>str</em>) – chromosome end inclusion type.
One of ‘inclusive’, ‘exclusive’ (default: ‘inclusive’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An Alignment instance.</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ValueError in case field 'score'</strong> – </p></li>
<li><p><strong>is not found in alignment fields.</strong> – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.Alignment.get_all_transcripts">
<code class="sig-name descname">get_all_transcripts</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.Alignment.get_all_transcripts" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds all possible transcripts from HSPs.</p>
<p>Finds all possible transcripts consisting of
consecutive non-overlapping HSPs in both
query and subject sequences in two orientations:
* direct, i.e. HSP is aimed from start towards</p>
<blockquote>
<div><p>end in both query and subject sequences;</p>
</div></blockquote>
<ul class="simple">
<li><dl class="simple">
<dt>reverse, i.e. from start to end in query</dt><dd><p>sequence but from end to start
in subject sequence.</p>
</dd>
</dl>
</li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Very greedy – exponential on number of HSPs.
Use only with few HSPs.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dictionary containing lists of transcripts
in direct and reverse orientation.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.Alignment.get_best_transcripts">
<code class="sig-name descname">get_best_transcripts</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.Alignment.get_best_transcripts" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds best transcripts.</p>
<p>Finds best transcripts in direct
and reverse HSP orientation with
dynamic programming approach (i.e.
minimizing gap penalties between HSPs.)</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict with best direct and reverse
Transcript instance under ‘direct’
and ‘reverse’ keys, respectively.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.Alignment.plot_alignment">
<code class="sig-name descname">plot_alignment</code><span class="sig-paren">(</span><em class="sig-param">qleft=None</em>, <em class="sig-param">qright=None</em>, <em class="sig-param">sleft=None</em>, <em class="sig-param">sright=None</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.Alignment.plot_alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots alignment map.</p>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.alignment.Alignment.replace_dict">
<code class="sig-name descname">replace_dict</code><em class="property"> = {'q. end': 'qend', 'q. start': 'qstart', 'query length': 'qlen', 's. end': 'send', 's. start': 'sstart', 'subject length': 'slen'}</em><a class="headerlink" href="#ortho2align.alignment.Alignment.replace_dict" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.Alignment.reset_filter_by_score">
<code class="sig-name descname">reset_filter_by_score</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.Alignment.reset_filter_by_score" title="Permalink to this definition">¶</a></dt>
<dd><p>Resets HSP filtering by score.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.Alignment.set_zero">
<code class="sig-name descname">set_zero</code><span class="sig-paren">(</span><em class="sig-param">qzero</em>, <em class="sig-param">szero</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.Alignment.set_zero" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets new query and subject sequence beginnings.</p>
<p>Given coordinates of the current query and subject
sequence beginnnings in the new coordinates,
transforms all HSP coordinates accordingly. Returns
new Alignment instance, filtration and graph data
will be lost. Use before filtering and graph building.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>qzero</strong> (<em>int</em>) – the location of the current query
sequence beginning in the new coordinates.</p></li>
<li><p><strong>szero</strong> (<em>int</em>) – the location of the current subject
sequence beginning in the new coordinates.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Alignment instance with new HSP coordinates
and new qlen and slen.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.Alignment.to_dict">
<code class="sig-name descname">to_dict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.Alignment.to_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Transforms instance to dict representation.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="ortho2align.alignment.HSP">
<em class="property">class </em><code class="sig-prename descclassname">ortho2align.alignment.</code><code class="sig-name descname">HSP</code><span class="sig-paren">(</span><em class="sig-param">qstart</em>, <em class="sig-param">qend</em>, <em class="sig-param">sstart</em>, <em class="sig-param">send</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.HSP" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>High-scoring pair from BLAST alignment representing
an alignment block.</p>
<dl class="method">
<dt id="ortho2align.alignment.HSP.__eq__">
<code class="sig-name descname">__eq__</code><span class="sig-paren">(</span><em class="sig-param">other</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.HSP.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Equality magic method.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.HSP.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">qstart</em>, <em class="sig-param">qend</em>, <em class="sig-param">sstart</em>, <em class="sig-param">send</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.HSP.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initializes HSP class.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>qstart</strong> (<em>int</em>) – start of HSP in query sequence.</p></li>
<li><p><strong>qend</strong> (<em>int</em>) – end of HSP in query sequence.</p></li>
<li><p><strong>sstart</strong> (<em>int</em>) – start of HSP in subject sequence.</p></li>
<li><p><strong>send</strong> (<em>int</em>) – end of HSP in subject sequence.</p></li>
<li><p><strong>score</strong> (<em>int, float</em>) – alignment score of HSP.</p></li>
<li><p><strong>qchrom</strong> (<em>int, str</em>) – chromosome of query sequence
(default: None).</p></li>
<li><p><strong>schrom</strong> (<em>int, str</em>) – chromosome of subject sequence
(default: None).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.HSP._process_boundary">
<em class="property">static </em><code class="sig-name descname">_process_boundary</code><span class="sig-paren">(</span><em class="sig-param">hsp</em>, <em class="sig-param">point</em>, <em class="sig-param">type_</em>, <em class="sig-param">inplace=True</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.HSP._process_boundary" title="Permalink to this definition">¶</a></dt>
<dd><p>Cuts HSP to a certain boundary.</p>
<p>In case an HSP intersects a boundary, it is cut
and its score is reduced proportionally to the
share of the cut.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>hsp</strong> (<em>HSP</em>) – an HSP instance to modify;</p></li>
<li><p><strong>point</strong> (<em>int, float</em>) – a coordinate to be set
new boundary;</p></li>
<li><p><strong>type_</strong> (<em>str</em>) – a type of point. One of ‘qleft’,
‘qright’, ‘sleft’, ‘sright’;</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – whether to modify HSP in-place
or return a modified copy (default: True)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None if inplace, modified copy of hsp otherwise.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.HSP.copy">
<code class="sig-name descname">copy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.HSP.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a copy of an instance.</p>
<p>All dynamic data in HSPVertex will be lost!</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.HSP.distance">
<code class="sig-name descname">distance</code><span class="sig-paren">(</span><em class="sig-param">other</em>, <em class="sig-param">gapopen=5</em>, <em class="sig-param">gapextend=2</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.HSP.distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates distance between two HSPs.</p>
<p>Distance between two HSPs of same orientation
is calculated as sum of distances in query and
subject sequence multiplied by gap extension
penalty added with gap opening penalty and
subtracted with the score of the following HSP.
In case none HSP precedes the other one (overlap
or different, non-sequential placement) the
distance is set to infinity.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>self</strong> – HSP instance;</p></li>
<li><p><strong>other</strong> – HSP instance;</p></li>
<li><p><strong>gapopen</strong> (<em>int</em>) – penalty for gap opening;</p></li>
<li><p><strong>gapextend</strong> (<em>int</em>) – penalty for gap extension.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An int/float or Inf in case none HSP precedes
the other one.</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ValueError in case of inconsistent orientations.</strong> – </p></li>
<li><p><strong>Distance is defined only for HSPs of same</strong> – </p></li>
<li><p><strong>orientation</strong><strong> (</strong><strong>except for None orientation</strong><strong>)</strong><strong></strong> – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.HSP.from_dict">
<em class="property">classmethod </em><code class="sig-name descname">from_dict</code><span class="sig-paren">(</span><em class="sig-param">dict_</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.HSP.from_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Recovers from a dict representation.</p>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.alignment.HSP.orientation_dict">
<code class="sig-name descname">orientation_dict</code><em class="property"> = {(True, False): 'reverse', (True, True): 'direct'}</em><a class="headerlink" href="#ortho2align.alignment.HSP.orientation_dict" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.HSP.precede">
<code class="sig-name descname">precede</code><span class="sig-paren">(</span><em class="sig-param">other</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.HSP.precede" title="Permalink to this definition">¶</a></dt>
<dd><p>Infers HSP precedence.</p>
<p>Infers whether self HSP precedes other HSP
in both query and subject sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True if self precedes other, False otherwise.</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError in case of inconsistent orientations.</strong> – </p></li>
<li><p><strong>Precedence is defined only for HSPs of same</strong> – </p></li>
<li><p><strong>orientation</strong><strong> (</strong><strong>except for None orientation</strong><strong>)</strong><strong></strong> – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.HSP.to_dict">
<code class="sig-name descname">to_dict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.HSP.to_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dict representation.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="ortho2align.alignment.HSPVertex">
<em class="property">class </em><code class="sig-prename descclassname">ortho2align.alignment.</code><code class="sig-name descname">HSPVertex</code><span class="sig-paren">(</span><em class="sig-param">qstart</em>, <em class="sig-param">qend</em>, <em class="sig-param">sstart</em>, <em class="sig-param">send</em>, <em class="sig-param">score</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.HSPVertex" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#ortho2align.alignment.HSP" title="ortho2align.alignment.HSP"><code class="xref py py-class docutils literal notranslate"><span class="pre">ortho2align.alignment.HSP</span></code></a></p>
<p>An extension of HSP class for handling graph
operations (contains a list of neighbours).</p>
<dl class="method">
<dt id="ortho2align.alignment.HSPVertex.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">qstart</em>, <em class="sig-param">qend</em>, <em class="sig-param">sstart</em>, <em class="sig-param">send</em>, <em class="sig-param">score</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.HSPVertex.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Inits HSPVertex class.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>qstart</strong> (<em>int</em>) – start of HSP in query sequence.</p></li>
<li><p><strong>qend</strong> (<em>int</em>) – end of HSP in query sequence.</p></li>
<li><p><strong>sstart</strong> (<em>int</em>) – start of HSP in subject sequence.</p></li>
<li><p><strong>send</strong> (<em>int</em>) – end of HSP in subject sequence.</p></li>
<li><p><strong>qchrom</strong> (<em>int, str</em>) – chromosome of query sequence
(default: None).</p></li>
<li><p><strong>schrom</strong> (<em>int, str</em>) – chromosome of subject sequence
(default: None).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.HSPVertex._relax">
<code class="sig-name descname">_relax</code><span class="sig-paren">(</span><em class="sig-param">other</em>, <em class="sig-param">weight</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.HSPVertex._relax" title="Permalink to this definition">¶</a></dt>
<dd><p>A relax method for dynamic programming.</p>
<p>If the current path from the begin vertex
to self vertex is more expensive than the path
that goes through the other vertex, a new cheaper
path from the begin to self is set.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>self</strong> (<em>HSPVertex</em>) – a vertex to be relaxed,
i.e. to minimize score of path from
begin vertex to the vertex.</p></li>
<li><p><strong>other</strong> (<em>HSPVertex</em>) – a vertex preceding
and connected to self.</p></li>
<li><p><strong>weight</strong> (<em>int, float</em>) – weight of the edge
connecting other and self vertices.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="ortho2align.alignment.Transcript">
<em class="property">class </em><code class="sig-prename descclassname">ortho2align.alignment.</code><code class="sig-name descname">Transcript</code><span class="sig-paren">(</span><em class="sig-param">HSPs</em>, <em class="sig-param">alignment</em>, <em class="sig-param">score=None</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.Transcript" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Transcript combined from non-overlapping
sequential HSPs.</p>
<dl class="attribute">
<dt id="ortho2align.alignment.Transcript.HSPs">
<code class="sig-name descname">HSPs</code><a class="headerlink" href="#ortho2align.alignment.Transcript.HSPs" title="Permalink to this definition">¶</a></dt>
<dd><p>a list of HSPVertex instances
representing a sequence of homologous
regions in query and subject sequences;</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.alignment.Transcript.alignment">
<code class="sig-name descname">alignment</code><a class="headerlink" href="#ortho2align.alignment.Transcript.alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>corresponding
<cite>Alignment</cite> instance;</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#ortho2align.alignment.Alignment" title="ortho2align.alignment.Alignment">Alignment</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.alignment.Transcript.score">
<code class="sig-name descname">score</code><a class="headerlink" href="#ortho2align.alignment.Transcript.score" title="Permalink to this definition">¶</a></dt>
<dd><p>a transcript score
computed as HSPs scores minus gap
penalties (see <cite>set_score</cite> and
<cite>HSP.distance</cite> methods for definition).</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int, float</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.Transcript.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">HSPs</em>, <em class="sig-param">alignment</em>, <em class="sig-param">score=None</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.Transcript.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Inits Transcript class.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>HSPs</strong> (<em>list</em>) – a list of HSPs representing the
transcript.</p></li>
<li><p><strong>alignment</strong> (<em>GenomicRangesAlignment</em>) – corresponding <cite>Alignment</cite>
instance.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.Transcript.from_dict">
<em class="property">classmethod </em><code class="sig-name descname">from_dict</code><span class="sig-paren">(</span><em class="sig-param">dict_</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.Transcript.from_dict" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.Transcript.plot_transcript">
<code class="sig-name descname">plot_transcript</code><span class="sig-paren">(</span><em class="sig-param">color='red'</em>, <em class="sig-param">link_color='blue'</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.Transcript.plot_transcript" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots transcript with corresponding alignment.</p>
<p>First plots corresponding alignment in black, then
transcripts’s HSPs with solid lines and connections
between them with dashed lines.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>color</strong> (<em>str</em>) – a color name to plot transcript’s
HSPs;</p></li>
<li><p><strong>link_color</strong> (<em>str</em>) – a color name to plot connections
between transcript’s HSPs.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt>
<em class="property">property </em><code class="sig-name descname">score</code></dt>
<dd><p>Calculates transcript score.</p>
<p>The transcript score is based on alignment
score and gap penalties defined by distance
function.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Transcript score.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.alignment.Transcript.to_dict">
<code class="sig-name descname">to_dict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.Transcript.to_dict" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="function">
<dt id="ortho2align.alignment.compare">
<code class="sig-prename descclassname">ortho2align.alignment.</code><code class="sig-name descname">compare</code><span class="sig-paren">(</span><em class="sig-param">a</em>, <em class="sig-param">b</em>, <em class="sig-param">side='g'</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.compare" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates boolean function of two values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>a</strong> – the first argument to compare
with the second argument.</p></li>
<li><p><strong>b</strong> – the second argument to compare
with the first argument.</p></li>
<li><p><strong>side</strong> (<em>str</em>) – boolean function code which
is used to compare a with b. The accepted
values are:
‘g’: greater, a &gt; b
‘l’: less, a &lt; b
‘eq’: equal, a == b
‘neq’: not equal, a != b
‘geq’: greater or equal, a &gt;= b
‘leq’: less or equal, a &lt;= b</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(bool) True or False, result of comparison
of a and b with side function.</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><strong>ValueError in case side is not one of 'g'</strong><strong>,</strong> – ‘l’, ‘eq’, ‘neq’, ‘geq’, ‘leq’.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="ortho2align.alignment.is_float">
<code class="sig-prename descclassname">ortho2align.alignment.</code><code class="sig-name descname">is_float</code><span class="sig-paren">(</span><em class="sig-param">string</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.is_float" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks whether a string represent a float number.</p>
</dd></dl>

<dl class="function">
<dt id="ortho2align.alignment.numberize">
<code class="sig-prename descclassname">ortho2align.alignment.</code><code class="sig-name descname">numberize</code><span class="sig-paren">(</span><em class="sig-param">string</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.numberize" title="Permalink to this definition">¶</a></dt>
<dd><p>String to number if possible.</p>
<p>Transforms string to float or to
integer if possible.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>int, float or str instance.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="ortho2align.alignment.nxor_strands">
<code class="sig-prename descclassname">ortho2align.alignment.</code><code class="sig-name descname">nxor_strands</code><span class="sig-paren">(</span><em class="sig-param">hsp_strand</em>, <em class="sig-param">alignment_strand</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.alignment.nxor_strands" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns NXOR of hsp strand and alignment strand.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>hsp_strand</strong> (<em>boolean</em>) – HSP.qstrand or HSP.sstrand.</p></li>
<li><p><strong>qlignment_strand</strong> (<em>str</em>) – Alignment.qstrand or Alignment.sstrand.
One of “+”, “-“, “.”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>“+” if both strands are concordant, “-” otherwise.</dt><dd><p>”.” is equal to “+”.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(str)</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-ortho2align.cli_scripts">
<span id="ortho2align-cli-scripts-module"></span><h2>ortho2align.cli_scripts module<a class="headerlink" href="#module-ortho2align.cli_scripts" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="ortho2align.cli_scripts.CLIVerbose">
<em class="property">class </em><code class="sig-prename descclassname">ortho2align.cli_scripts.</code><code class="sig-name descname">CLIVerbose</code><span class="sig-paren">(</span><em class="sig-param">messages</em>, <em class="sig-param">target=&lt;_io.TextIOWrapper name='&lt;stdout&gt;' mode='w' encoding='UTF-8'&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.cli_scripts.CLIVerbose" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A context manager for verbosity before and after execution
of some commands.</p>
<dl class="method">
<dt id="ortho2align.cli_scripts.CLIVerbose.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">messages</em>, <em class="sig-param">target=&lt;_io.TextIOWrapper name='&lt;stdout&gt;' mode='w' encoding='UTF-8'&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.cli_scripts.CLIVerbose.__init__" title="Permalink to this definition">¶</a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>messages</strong> (<em>list</em>) – list with 2 items containing start message
and finish message.</p></li>
<li><p><strong>target</strong> (<em>fp</em>) – where to print messages (default: sys.stdout).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="ortho2align.cli_scripts.align_syntenies">
<code class="sig-prename descclassname">ortho2align.cli_scripts.</code><code class="sig-name descname">align_syntenies</code><span class="sig-paren">(</span><em class="sig-param">grange</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.cli_scripts.align_syntenies" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="ortho2align.cli_scripts.align_two_ranges">
<code class="sig-prename descclassname">ortho2align.cli_scripts.</code><code class="sig-name descname">align_two_ranges</code><span class="sig-paren">(</span><em class="sig-param">query</em>, <em class="sig-param">subject</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.cli_scripts.align_two_ranges" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="ortho2align.cli_scripts.cache_orthodb_xrefs">
<code class="sig-prename descclassname">ortho2align.cli_scripts.</code><code class="sig-name descname">cache_orthodb_xrefs</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.cli_scripts.cache_orthodb_xrefs" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="ortho2align.cli_scripts.estimate_background">
<code class="sig-prename descclassname">ortho2align.cli_scripts.</code><code class="sig-name descname">estimate_background</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.cli_scripts.estimate_background" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="ortho2align.cli_scripts.get_alignments">
<code class="sig-prename descclassname">ortho2align.cli_scripts.</code><code class="sig-name descname">get_alignments</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.cli_scripts.get_alignments" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="ortho2align.cli_scripts.get_orthodb_map">
<code class="sig-prename descclassname">ortho2align.cli_scripts.</code><code class="sig-name descname">get_orthodb_map</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.cli_scripts.get_orthodb_map" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="ortho2align.cli_scripts.ortho2align">
<code class="sig-prename descclassname">ortho2align.cli_scripts.</code><code class="sig-name descname">ortho2align</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.cli_scripts.ortho2align" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="ortho2align.cli_scripts.refine_alignments">
<code class="sig-prename descclassname">ortho2align.cli_scripts.</code><code class="sig-name descname">refine_alignments</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.cli_scripts.refine_alignments" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="module-ortho2align.fitting">
<span id="ortho2align-fitting-module"></span><h2>ortho2align.fitting module<a class="headerlink" href="#module-ortho2align.fitting" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="ortho2align.fitting.AbstractFitter">
<em class="property">class </em><code class="sig-prename descclassname">ortho2align.fitting.</code><code class="sig-name descname">AbstractFitter</code><a class="headerlink" href="#ortho2align.fitting.AbstractFitter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Abstract class for empirical distribution estimation.</p>
<dl class="method">
<dt id="ortho2align.fitting.AbstractFitter.__subclasshook__">
<em class="property">classmethod </em><code class="sig-name descname">__subclasshook__</code><span class="sig-paren">(</span><em class="sig-param">C</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.fitting.AbstractFitter.__subclasshook__" title="Permalink to this definition">¶</a></dt>
<dd><p>Magic method for issubclass() function.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.fitting.AbstractFitter.cdf">
<em class="property">abstract </em><code class="sig-name descname">cdf</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.fitting.AbstractFitter.cdf" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns cumulative density function of the items in data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>number or iterable</em>) – data to estiamate
cdf with fitted empirical distribution.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(number or np.array) cdf values of the items in data.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.fitting.AbstractFitter.estimator">
<em class="property">abstract property </em><code class="sig-name descname">estimator</code><a class="headerlink" href="#ortho2align.fitting.AbstractFitter.estimator" title="Permalink to this definition">¶</a></dt>
<dd><p>Abstract property for estimator object.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.fitting.AbstractFitter.isf">
<em class="property">abstract </em><code class="sig-name descname">isf</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.fitting.AbstractFitter.isf" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns inverse survival function of the items in data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>number or iterable</em>) – data to estiamate
isf with fitted empirical distribution.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(number or np.array) isf values of the items in data.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.fitting.AbstractFitter.pdf">
<em class="property">abstract </em><code class="sig-name descname">pdf</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.fitting.AbstractFitter.pdf" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns probability density function of the items in data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>number or iterable</em>) – data to estiamate
pdf with fitted empirical distribution.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(number or np.array) pdf values of the items in data.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.fitting.AbstractFitter.ppf">
<em class="property">abstract </em><code class="sig-name descname">ppf</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.fitting.AbstractFitter.ppf" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns percent point function of the items in data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>number or iterable</em>) – data to estiamate
ppf with fitted empirical distribution.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(number or np.array) ppf values of the items in data.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.fitting.AbstractFitter.sf">
<em class="property">abstract </em><code class="sig-name descname">sf</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.fitting.AbstractFitter.sf" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns survival function of the items in data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>number or iterable</em>) – data to estiamate
sf with fitted empirical distribution.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(number or np.array) sf values of the items in data.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="ortho2align.fitting.HistogramFitter">
<em class="property">class </em><code class="sig-prename descclassname">ortho2align.fitting.</code><code class="sig-name descname">HistogramFitter</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.fitting.HistogramFitter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#ortho2align.fitting.AbstractFitter" title="ortho2align.fitting.AbstractFitter"><code class="xref py py-class docutils literal notranslate"><span class="pre">ortho2align.fitting.AbstractFitter</span></code></a></p>
<p>Fits empirical data with histogram and produces continuous distribution.</p>
<dl class="attribute">
<dt id="ortho2align.fitting.HistogramFitter.data">
<code class="sig-name descname">data</code><a class="headerlink" href="#ortho2align.fitting.HistogramFitter.data" title="Permalink to this definition">¶</a></dt>
<dd><p>numerical data used to fit distribution.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>collection</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.fitting.HistogramFitter.estimator">
<code class="sig-name descname">estimator</code><a class="headerlink" href="#ortho2align.fitting.HistogramFitter.estimator" title="Permalink to this definition">¶</a></dt>
<dd><p>statistics.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>scipy.rv_histogram</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.fitting.HistogramFitter.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.fitting.HistogramFitter.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initializes a HistogramFitter instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>iterable</em>) – empirical data to be fitted.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.fitting.HistogramFitter.cdf">
<code class="sig-name descname">cdf</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.fitting.HistogramFitter.cdf" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns cumulative density function of the items in data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>number or iterable</em>) – data to estiamate
cdf with fitted empirical distribution.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(number or np.array) cdf values of the items in data.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt>
<em class="property">property </em><code class="sig-name descname">estimator</code></dt>
<dd><p>Returns _estimator property.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.fitting.HistogramFitter.isf">
<code class="sig-name descname">isf</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.fitting.HistogramFitter.isf" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns inverse survival function of the items in data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>number or iterable</em>) – data to estiamate
isf with fitted empirical distribution.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(number or np.array) isf values of the items in data.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.fitting.HistogramFitter.pdf">
<code class="sig-name descname">pdf</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.fitting.HistogramFitter.pdf" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns probability density function of the items in data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>number or iterable</em>) – data to estiamate
pdf with fitted empirical distribution.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(number or np.array) pdf values of the items in data.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.fitting.HistogramFitter.ppf">
<code class="sig-name descname">ppf</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.fitting.HistogramFitter.ppf" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns percent point function of the items in data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>number or iterable</em>) – data to estiamate
ppf with fitted empirical distribution.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(number or np.array) ppf values of the items in data.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.fitting.HistogramFitter.sf">
<code class="sig-name descname">sf</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.fitting.HistogramFitter.sf" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns survival function of the items in data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>number or iterable</em>) – data to estiamate
sf with fitted empirical distribution.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(number or np.array) sf values of the items in data.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="ortho2align.fitting.KernelFitter">
<em class="property">class </em><code class="sig-prename descclassname">ortho2align.fitting.</code><code class="sig-name descname">KernelFitter</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.fitting.KernelFitter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#ortho2align.fitting.AbstractFitter" title="ortho2align.fitting.AbstractFitter"><code class="xref py py-class docutils literal notranslate"><span class="pre">ortho2align.fitting.AbstractFitter</span></code></a></p>
<p>Fits empirical data with KDE and produces continuous distribution.</p>
<dl class="attribute">
<dt id="ortho2align.fitting.KernelFitter.data">
<code class="sig-name descname">data</code><a class="headerlink" href="#ortho2align.fitting.KernelFitter.data" title="Permalink to this definition">¶</a></dt>
<dd><p>numerical data used to fit distribution.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>collection</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.fitting.KernelFitter.estimator">
<code class="sig-name descname">estimator</code><a class="headerlink" href="#ortho2align.fitting.KernelFitter.estimator" title="Permalink to this definition">¶</a></dt>
<dd><p>statistics.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>scipy.gaussian_kde</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.fitting.KernelFitter.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.fitting.KernelFitter.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initializes KernelFitter instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>collection</em>) – numerical data to be fitted.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.fitting.KernelFitter.cdf">
<code class="sig-name descname">cdf</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.fitting.KernelFitter.cdf" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns cumulative density function of the items in data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>number or iterable</em>) – data to estiamate
cdf with fitted empirical distribution.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(number or np.array) cdf values of the items in data.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt>
<em class="property">property </em><code class="sig-name descname">estimator</code></dt>
<dd><p>Returns _estimator property.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.fitting.KernelFitter.isf">
<code class="sig-name descname">isf</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">epsilon='auto'</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.fitting.KernelFitter.isf" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns inverse survival function of the items in data.</p>
<p>Utilizes binary search to find isf of values in data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>number or iterable</em>) – data to estiamate
isf with fitted empirical distribution.</p></li>
<li><p><strong>epsilon</strong> (<em>int, float, str</em>) – the maximum difference between data and isf(points)
during the search to stop it. If ‘auto’, than selects average size
between data points used to fit distribution (default: ‘auto’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(number or np.array) isf values of the items in data.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.fitting.KernelFitter.pdf">
<code class="sig-name descname">pdf</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.fitting.KernelFitter.pdf" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns probability density function of the items in data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>number or iterable</em>) – data to estiamate
pdf with fitted empirical distribution.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(number or np.array) pdf values of the items in data.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.fitting.KernelFitter.ppf">
<code class="sig-name descname">ppf</code><span class="sig-paren">(</span><em class="sig-param">data</em>, <em class="sig-param">epsilon='auto'</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.fitting.KernelFitter.ppf" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns percent point function of the items in data.</p>
<p>Utilizes binary search to find ppf of values in data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>number or iterable</em>) – data to estiamate
ppf with fitted empirical distribution.</p></li>
<li><p><strong>epsilon</strong> (<em>int, float, str</em>) – the maximum difference between data and ppf(points)
during the search to stop it. If ‘auto’, than selects average size
between data points used to fit distribution (default: ‘auto’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(number or np.array) ppf values of the items in data.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.fitting.KernelFitter.sf">
<code class="sig-name descname">sf</code><span class="sig-paren">(</span><em class="sig-param">data</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.fitting.KernelFitter.sf" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns survival function of the items in data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>number or iterable</em>) – data to estiamate
sf with fitted empirical distribution.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(number or np.array) sf values of the items in data.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="ortho2align.fitting.inverse_approximator">
<code class="sig-prename descclassname">ortho2align.fitting.</code><code class="sig-name descname">inverse_approximator</code><span class="sig-paren">(</span><em class="sig-param">value</em>, <em class="sig-param">function</em>, <em class="sig-param">left</em>, <em class="sig-param">right</em>, <em class="sig-param">epsilon</em>, <em class="sig-param">increasing=True</em>, <em class="sig-param">iterations=100</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.fitting.inverse_approximator" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds argument of monotonic function given its result.</p>
<p>Utilizes binary search algorithm to find the best argument
so that the function result of it is the closest to the
providied value. Search range is provided; function must
be monotonic on this search range.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>value</strong> (<em>int, float</em>) – result of function.</p></li>
<li><p><strong>function</strong> (<em>callable</em>) – function to calculate value.</p></li>
<li><p><strong>left</strong> (<em>int, float</em>) – left boundary of the search range.</p></li>
<li><p><strong>right</strong> (<em>int, float</em>) – right boundary of the search range.</p></li>
<li><p><strong>epsilon</strong> (<em>int, float</em>) – the maximum difference between value
and calculated result of function during the search to
stop it.</p></li>
<li><p><strong>increasing</strong> (<em>bool</em>) – whether the function is increasing on the
search range or not (default: True).</p></li>
<li><p><strong>iterations</strong> (<em>int</em>) – number of search iterations to perform
until halt (default: 100).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(int, float) the best argument found.</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-ortho2align.genomicranges">
<span id="ortho2align-genomicranges-module"></span><h2>ortho2align.genomicranges module<a class="headerlink" href="#module-ortho2align.genomicranges" title="Permalink to this headline">¶</a></h2>
<p>Chromosome coordinate systems.
0-based: chromosome start coordinate is 0.
1-based: chromosome start coordinate is 1.
inclusive: character under end coordinate</p>
<blockquote>
<div><p>is included in the sequence.</p>
</div></blockquote>
<dl class="simple">
<dt>exclusive: character under end coordinate</dt><dd><p>is not included in the sequence.</p>
</dd>
</dl>
<p>Here we use 0-based exclusive system for
easier arithmetics.</p>
<p>Translation from other systems:
0-based inclusive: start += 0, end += 1
1-based inclusive: start += -1, end += 0
1-based exclusive: start += -1, end += -1</p>
<dl class="class">
<dt id="ortho2align.genomicranges.AlignedRangePair">
<em class="property">class </em><code class="sig-prename descclassname">ortho2align.genomicranges.</code><code class="sig-name descname">AlignedRangePair</code><span class="sig-paren">(</span><em class="sig-param">query_grange</em>, <em class="sig-param">subject_grange</em>, <em class="sig-param">alignment</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.AlignedRangePair" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>An association of genomic ranges and their alignment.</p>
<p>Stores query genomic range, subject genomic range and
their alignment.</p>
<dl class="attribute">
<dt id="ortho2align.genomicranges.AlignedRangePair.query_grange">
<code class="sig-name descname">query_grange</code><a class="headerlink" href="#ortho2align.genomicranges.AlignedRangePair.query_grange" title="Permalink to this definition">¶</a></dt>
<dd><p>query genomic range.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#ortho2align.genomicranges.GenomicRange" title="ortho2align.genomicranges.GenomicRange">GenomicRange</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.genomicranges.AlignedRangePair.subject_grange">
<code class="sig-name descname">subject_grange</code><a class="headerlink" href="#ortho2align.genomicranges.AlignedRangePair.subject_grange" title="Permalink to this definition">¶</a></dt>
<dd><p>subject genomic range.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#ortho2align.genomicranges.GenomicRange" title="ortho2align.genomicranges.GenomicRange">GenomicRange</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.genomicranges.AlignedRangePair.alignment">
<code class="sig-name descname">alignment</code><a class="headerlink" href="#ortho2align.genomicranges.AlignedRangePair.alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>alignment of query and subject</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#ortho2align.alignment.Alignment" title="ortho2align.alignment.Alignment">Alignment</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt>
<code class="sig-name descname">genomic ranges.</code></dt>
<dd></dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.AlignedRangePair.__eq__">
<code class="sig-name descname">__eq__</code><span class="sig-paren">(</span><em class="sig-param">other</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.AlignedRangePair.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Equality magic method.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.AlignedRangePair.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">query_grange</em>, <em class="sig-param">subject_grange</em>, <em class="sig-param">alignment</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.AlignedRangePair.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initializes AlignedRangePair instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>query_grange</strong> (<em>GenomicRange</em>) – query genomic range.</p></li>
<li><p><strong>subject_grange</strong> (<em>GenomicRange</em>) – subject genomic range.</p></li>
<li><p><strong>alignment</strong> (<em>Alignment</em>) – alignment of query and subject</p></li>
<li><p><strong>genomic ranges.</strong></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.AlignedRangePair.__repr__">
<code class="sig-name descname">__repr__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.AlignedRangePair.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns representation of AlignedRangePair instance.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.AlignedRangePair.__str__">
<code class="sig-name descname">__str__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.AlignedRangePair.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns string representation.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.AlignedRangePair.from_dict">
<em class="property">classmethod </em><code class="sig-name descname">from_dict</code><span class="sig-paren">(</span><em class="sig-param">dict_</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.AlignedRangePair.from_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Restores AlignedRangePair instance from dict representation.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.AlignedRangePair.to_dict">
<code class="sig-name descname">to_dict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.AlignedRangePair.to_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns dict representation of AlignedRangePair instance.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="ortho2align.genomicranges.ChromosomeLocation">
<em class="property">class </em><code class="sig-prename descclassname">ortho2align.genomicranges.</code><code class="sig-name descname">ChromosomeLocation</code><span class="sig-paren">(</span><em class="sig-param">size</em>, <em class="sig-param">start</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.ChromosomeLocation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code></p>
<dl class="method">
<dt id="ortho2align.genomicranges.ChromosomeLocation.size">
<em class="property">property </em><code class="sig-name descname">size</code><a class="headerlink" href="#ortho2align.genomicranges.ChromosomeLocation.size" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 0</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.ChromosomeLocation.start">
<em class="property">property </em><code class="sig-name descname">start</code><a class="headerlink" href="#ortho2align.genomicranges.ChromosomeLocation.start" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 1</p>
</dd></dl>

</dd></dl>

<dl class="exception">
<dt id="ortho2align.genomicranges.ChromosomeNotFoundError">
<em class="property">exception </em><code class="sig-prename descclassname">ortho2align.genomicranges.</code><code class="sig-name descname">ChromosomeNotFoundError</code><span class="sig-paren">(</span><em class="sig-param">chrom</em>, <em class="sig-param">genome</em>, <em class="sig-param">list_of_chroms</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.ChromosomeNotFoundError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#ortho2align.genomicranges.GenomicException" title="ortho2align.genomicranges.GenomicException"><code class="xref py py-class docutils literal notranslate"><span class="pre">ortho2align.genomicranges.GenomicException</span></code></a></p>
<p>Occurs when chromosome was not found in given genome.</p>
</dd></dl>

<dl class="exception">
<dt id="ortho2align.genomicranges.EmptyGenomicRangesListError">
<em class="property">exception </em><code class="sig-prename descclassname">ortho2align.genomicranges.</code><code class="sig-name descname">EmptyGenomicRangesListError</code><span class="sig-paren">(</span><em class="sig-param">range_list_instance</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.EmptyGenomicRangesListError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#ortho2align.genomicranges.GenomicException" title="ortho2align.genomicranges.GenomicException"><code class="xref py py-class docutils literal notranslate"><span class="pre">ortho2align.genomicranges.GenomicException</span></code></a></p>
<p>Occurs when GenomicRangesList is empty.</p>
</dd></dl>

<dl class="class">
<dt id="ortho2align.genomicranges.FastaSeqFile">
<em class="property">class </em><code class="sig-prename descclassname">ortho2align.genomicranges.</code><code class="sig-name descname">FastaSeqFile</code><span class="sig-paren">(</span><em class="sig-param">sequence_file_path</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.FastaSeqFile" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Represents genomic fasta file.</p>
<p>Should locate to the fasta file
where the genome is placed. Chromosome
names must be placed in fasta record
header and separated from other header
information with whitespace character.</p>
<dl class="attribute">
<dt id="ortho2align.genomicranges.FastaSeqFile.sequence_file_path">
<code class="sig-name descname">sequence_file_path</code><a class="headerlink" href="#ortho2align.genomicranges.FastaSeqFile.sequence_file_path" title="Permalink to this definition">¶</a></dt>
<dd><p>path to the file.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>Path</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.genomicranges.FastaSeqFile.chromsizes">
<code class="sig-name descname">chromsizes</code><a class="headerlink" href="#ortho2align.genomicranges.FastaSeqFile.chromsizes" title="Permalink to this definition">¶</a></dt>
<dd><p>dictionary of
ChromosomseLocation instances.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.genomicranges.FastaSeqFile._file_obj">
<code class="sig-name descname">_file_obj</code><a class="headerlink" href="#ortho2align.genomicranges.FastaSeqFile._file_obj" title="Permalink to this definition">¶</a></dt>
<dd><p>opened file under the
sequence_file_path.</p>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.genomicranges.FastaSeqFile.line_length">
<code class="sig-name descname">line_length</code><a class="headerlink" href="#ortho2align.genomicranges.FastaSeqFile.line_length" title="Permalink to this definition">¶</a></dt>
<dd><p>line length of fasta sequence.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.FastaSeqFile.__enter__">
<code class="sig-name descname">__enter__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.FastaSeqFile.__enter__" title="Permalink to this definition">¶</a></dt>
<dd><p>Context manager for opening sequence file.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.FastaSeqFile.__eq__">
<code class="sig-name descname">__eq__</code><span class="sig-paren">(</span><em class="sig-param">other</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.FastaSeqFile.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Equality magic method.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.FastaSeqFile.__exit__">
<code class="sig-name descname">__exit__</code><span class="sig-paren">(</span><em class="sig-param">exc_type</em>, <em class="sig-param">exc_val</em>, <em class="sig-param">exc_tb</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.FastaSeqFile.__exit__" title="Permalink to this definition">¶</a></dt>
<dd><p>Context manager for closing sequence file.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.FastaSeqFile.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">sequence_file_path</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.FastaSeqFile.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initializes FastaSeqFile instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sequence_file_path</strong> (<em>str, Path</em>) – path to the file.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.FastaSeqFile.__repr__">
<code class="sig-name descname">__repr__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.FastaSeqFile.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns representation of FastaSeqFile object.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.FastaSeqFile.__str__">
<code class="sig-name descname">__str__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.FastaSeqFile.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns string representation.</p>
</dd></dl>

<dl class="method">
<dt>
<em class="property">property </em><code class="sig-name descname">chromsizes</code></dt>
<dd><p>Property representing genomic chromosomes.</p>
<p>Calculates chromosome names, their sizes and
sequence start positions in the genome file.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><dl class="simple">
<dt>(dict) dictionary of ChromosomeLocation</dt><dd><p>instances in the following scheme:
{chromosome_name: ChromosomeLocation(size, start)}</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>FilePathNotSpecifiedError in case the location of</strong> – sequence file is not specified.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.FastaSeqFile.from_dict">
<em class="property">classmethod </em><code class="sig-name descname">from_dict</code><span class="sig-paren">(</span><em class="sig-param">dict_</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.FastaSeqFile.from_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Restores FastaSeqFile instance from its dict representation.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.FastaSeqFile.get_fasta_by_coord">
<code class="sig-name descname">get_fasta_by_coord</code><span class="sig-paren">(</span><em class="sig-param">grange</em>, <em class="sig-param">outfile</em>, <em class="sig-param">linewidth=60</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.FastaSeqFile.get_fasta_by_coord" title="Permalink to this definition">¶</a></dt>
<dd><p>Writes sequence of the genomic range in fasta file.</p>
<blockquote>
<div><p>Writes full fasta record of the given genomic range
to the file. The header is made in the following
scheme:
“&gt;” + grange.sequence_header + “</p>
</div></blockquote>
<p>“</p>
<blockquote>
<div><dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>grange (GenomicRange): genomic range which fasta</dt><dd><p>sequence write to file.</p>
</dd>
<dt>outfile (fileobj): file object to which write fasta</dt><dd><p>sequence.</p>
</dd>
<dt>linewidth (int): line length of fasta sequence to write</dt><dd><p>(default: 60).</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>None.</p>
</dd>
<dt>Raises:</dt><dd><dl class="simple">
<dt>ChromosomeNotFoundError if the chromosome of</dt><dd><p>the given genomic range is not presented
in FastaSeqFile instance.</p>
</dd>
</dl>
</dd>
</dl>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt>
<em class="property">property </em><code class="sig-name descname">line_length</code></dt>
<dd><p><cite>line_length</cite> getter.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>(int) length of sequence line in fasta file.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.FastaSeqFile.locate_coordinate">
<code class="sig-name descname">locate_coordinate</code><span class="sig-paren">(</span><em class="sig-param">chrom_start</em>, <em class="sig-param">coord</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.FastaSeqFile.locate_coordinate" title="Permalink to this definition">¶</a></dt>
<dd><p>Puts file pointer to the beginning of the genomic range.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chrom_start</strong> (<em>int</em>) – byte number of chromosome start.</p></li>
<li><p><strong>coord</strong> (<em>int</em>) – coordinate in the corresponding chromosome.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt>
<em class="property">property </em><code class="sig-name descname">sequence_file_path</code></dt>
<dd><p><cite>sequence_file_path</cite> getter</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.FastaSeqFile.to_dict">
<code class="sig-name descname">to_dict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.FastaSeqFile.to_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns dict representation.</p>
</dd></dl>

</dd></dl>

<dl class="exception">
<dt id="ortho2align.genomicranges.FilePathNotSpecifiedError">
<em class="property">exception </em><code class="sig-prename descclassname">ortho2align.genomicranges.</code><code class="sig-name descname">FilePathNotSpecifiedError</code><span class="sig-paren">(</span><em class="sig-param">instance</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.FilePathNotSpecifiedError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#ortho2align.genomicranges.GenomicException" title="ortho2align.genomicranges.GenomicException"><code class="xref py py-class docutils literal notranslate"><span class="pre">ortho2align.genomicranges.GenomicException</span></code></a></p>
<p>Occurs when location of sequence file is not specified, but is accessed.</p>
</dd></dl>

<dl class="exception">
<dt id="ortho2align.genomicranges.GenomicCoordinatesError">
<em class="property">exception </em><code class="sig-prename descclassname">ortho2align.genomicranges.</code><code class="sig-name descname">GenomicCoordinatesError</code><span class="sig-paren">(</span><em class="sig-param">grange</em>, <em class="sig-param">chromosome</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicCoordinatesError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#ortho2align.genomicranges.GenomicException" title="ortho2align.genomicranges.GenomicException"><code class="xref py py-class docutils literal notranslate"><span class="pre">ortho2align.genomicranges.GenomicException</span></code></a></p>
<p>Occurs when coordinates of GenomicRange instance are out of the chromosome.</p>
</dd></dl>

<dl class="exception">
<dt id="ortho2align.genomicranges.GenomicException">
<em class="property">exception </em><code class="sig-prename descclassname">ortho2align.genomicranges.</code><code class="sig-name descname">GenomicException</code><a class="headerlink" href="#ortho2align.genomicranges.GenomicException" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Exception</span></code></p>
<p>Basic exception in genomicranges module.</p>
</dd></dl>

<dl class="class">
<dt id="ortho2align.genomicranges.GenomicRange">
<em class="property">class </em><code class="sig-prename descclassname">ortho2align.genomicranges.</code><code class="sig-name descname">GenomicRange</code><span class="sig-paren">(</span><em class="sig-param">chrom</em>, <em class="sig-param">start</em>, <em class="sig-param">end</em>, <em class="sig-param">strand='.'</em>, <em class="sig-param">name=None</em>, <em class="sig-param">genome=None</em>, <em class="sig-param">sequence_file_path=None</em>, <em class="sig-param">relations=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRange" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Represents one genomic range.</p>
<p>Genomic coordinates are 0-based exclusive:
* chromosome start coordinate is 0;
* character under genomic range coordinate</p>
<blockquote>
<div><p>is not included into the sequence.</p>
</div></blockquote>
<dl class="attribute">
<dt id="ortho2align.genomicranges.GenomicRange.chrom">
<code class="sig-name descname">chrom</code><a class="headerlink" href="#ortho2align.genomicranges.GenomicRange.chrom" title="Permalink to this definition">¶</a></dt>
<dd><p>name of chromosome.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.genomicranges.GenomicRange.start">
<code class="sig-name descname">start</code><a class="headerlink" href="#ortho2align.genomicranges.GenomicRange.start" title="Permalink to this definition">¶</a></dt>
<dd><p>start of genomic range.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.genomicranges.GenomicRange.end">
<code class="sig-name descname">end</code><a class="headerlink" href="#ortho2align.genomicranges.GenomicRange.end" title="Permalink to this definition">¶</a></dt>
<dd><p>end of genomic range.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.genomicranges.GenomicRange.strand">
<code class="sig-name descname">strand</code><a class="headerlink" href="#ortho2align.genomicranges.GenomicRange.strand" title="Permalink to this definition">¶</a></dt>
<dd><p>genomic range strand. One of
“+”, “-“, “.” (inessential strandness)
(default: None).</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.genomicranges.GenomicRange.name">
<code class="sig-name descname">name</code><a class="headerlink" href="#ortho2align.genomicranges.GenomicRange.name" title="Permalink to this definition">¶</a></dt>
<dd><p>name of genomic range (defalut: None).</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.genomicranges.GenomicRange.genome">
<code class="sig-name descname">genome</code><a class="headerlink" href="#ortho2align.genomicranges.GenomicRange.genome" title="Permalink to this definition">¶</a></dt>
<dd><p>corresponding genome filename.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.genomicranges.GenomicRange.sequence_file_path">
<code class="sig-name descname">sequence_file_path</code><a class="headerlink" href="#ortho2align.genomicranges.GenomicRange.sequence_file_path" title="Permalink to this definition">¶</a></dt>
<dd><p>location of genomic range
sequence file.</p>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.genomicranges.GenomicRange.relations">
<code class="sig-name descname">relations</code><a class="headerlink" href="#ortho2align.genomicranges.GenomicRange.relations" title="Permalink to this definition">¶</a></dt>
<dd><p>a dictionary with related GenomicRangesList
instances.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRange.__eq__">
<code class="sig-name descname">__eq__</code><span class="sig-paren">(</span><em class="sig-param">other</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRange.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Equality magic method.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRange.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">chrom</em>, <em class="sig-param">start</em>, <em class="sig-param">end</em>, <em class="sig-param">strand='.'</em>, <em class="sig-param">name=None</em>, <em class="sig-param">genome=None</em>, <em class="sig-param">sequence_file_path=None</em>, <em class="sig-param">relations=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRange.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initializes GenomicRange instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chrom</strong> (<em>str</em>) – name of chromosome.</p></li>
<li><p><strong>start</strong> (<em>int</em>) – start of genomic range.</p></li>
<li><p><strong>end</strong> (<em>int</em>) – end of genomic range.</p></li>
<li><p><strong>strand</strong> (<em>str</em>) – genomic range strand. One of
“+”, “-“, “.” (inessential strandness)
(default: None).</p></li>
<li><p><strong>name</strong> (<em>str</em>) – name of genomic range (defalut: None).</p></li>
<li><p><strong>genome</strong> (<em>str</em>) – corresponding genome filename.</p></li>
<li><p><strong>sequence_file_path</strong> – location of genomic range
sequence file.</p></li>
<li><p><strong>relations</strong> (<em>dict</em>) – a dictionary with related GenomicRangesList
instances.</p></li>
<li><p><strong>kwargs</strong> (<em>dict</em>) – other arguments.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRange.__repr__">
<code class="sig-name descname">__repr__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRange.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>GenomicRange instance representation.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRange.__str__">
<code class="sig-name descname">__str__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRange.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>GenomicRange instance string representation.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRange.align_blast">
<code class="sig-name descname">align_blast</code><span class="sig-paren">(</span><em class="sig-param">other</em>, <em class="sig-param">program='blastn'</em>, <em class="sig-param">task='blastn'</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRange.align_blast" title="Permalink to this definition">¶</a></dt>
<dd><p>Aligns two genomic ranges with BLAST.</p>
<p>If both genomic ranges have their sequences
stored in separate files, aligns them with
standalone blastn and specified parameters.
Alignment file format is “7 std score”.
Standalone blast package should be installed
manually.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<em>GenomicRange</em>) – genomic range to
align self with.</p></li>
<li><p><strong>program</strong> (<em>str</em>) – a program from Standalone
BLAST to use (default: ‘blastn’)</p></li>
<li><p><strong>task</strong> (<em>str</em>) – which task from a BLAST program
use (default: ‘blastn’)</p></li>
<li><p><strong>kwargs</strong> (<em>dict</em>) – any command line arguments
to blastn in the following scheme:
{‘flag’: ‘value’}.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>aligned range pair,</dt><dd><p>which associates self and other
genomic ranges and their alignment.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(<a class="reference internal" href="#ortho2align.genomicranges.GenomicRangesAlignment" title="ortho2align.genomicranges.GenomicRangesAlignment">GenomicRangesAlignment</a>)</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>FilePathNotSpecifiedError in case one</strong><strong> or </strong><strong>both</strong> – </p></li>
<li><p><strong>genomic ranges do not have separate</strong> – </p></li>
<li><p><strong>sequence files.</strong> – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRange.align_with_relation">
<code class="sig-name descname">align_with_relation</code><span class="sig-paren">(</span><em class="sig-param">relation</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRange.align_with_relation" title="Permalink to this definition">¶</a></dt>
<dd><p>Aligns genomic range with all its relations.</p>
<p>Aligns given genomic range with all its
relations using standalone blastn
and returns list of GenomicRangesAlignment
instances.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>relation</strong> (<em>str</em>) – relation type.</p></li>
<li><p><strong>kwargs</strong> (<em>dict</em>) – any command line arguments
to blastn in the following scheme:
{‘flag’: ‘value’}.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of GenomicRangesAlignment instances.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(list)</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError in case provided relation is not</strong> – available for the genomic range.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRange.distance">
<code class="sig-name descname">distance</code><span class="sig-paren">(</span><em class="sig-param">other</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRange.distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates distance between two genomic ranges.</p>
<p>In case of same genomes and chromosomes
the distance between two genomic ranges
can be calculated regardless of their
strandness. If two genomic ranges overlap,
the distance is zero. In case of inconsistent
chromosomes or genomes the distance is undefined.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>GenomicRange</em>) – the genomic range
to which calculate the distance from
self.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(int) distance, non-negative number.</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>InconsistentGenomesError in case of inconsistent</strong> – genomes.</p></li>
<li><p><strong>InconsistentChromosomesError in case of inconsistent</strong> – chromosomes.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRange.find_neighbours">
<code class="sig-name descname">find_neighbours</code><span class="sig-paren">(</span><em class="sig-param">other_list</em>, <em class="sig-param">distance=0</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRange.find_neighbours" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds neighbours of self in other list at given distance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other_list</strong> (<em>GenomicRangesList</em>) – other list to find
neighbours in.</p></li>
<li><p><strong>distance</strong> (<em>int, float</em>) – distance at which two
genomic ranges are considered neighbours
(default: 0).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(GenomicRangesList) genomic ranges list with all
neighbours of self in other_list at given distance.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRange.flank">
<code class="sig-name descname">flank</code><span class="sig-paren">(</span><em class="sig-param">flank_distance=0</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRange.flank" title="Permalink to this definition">¶</a></dt>
<dd><p>Flanks genomic range with the given distance.</p>
<p>Start coordinate is decreased and end
coordinate is increased by the flank distance.
Do not check if the start coordinate is negative or
the end coordinate exceeds chromosome size.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>flank_distance</strong> (<em>int</em>) – distance to flank
genomic range.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(GenomicRange) flanked genomic range.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRange.from_dict">
<em class="property">classmethod </em><code class="sig-name descname">from_dict</code><span class="sig-paren">(</span><em class="sig-param">dict_</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRange.from_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Restores genomic range from dict representation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>dict_</strong> – dict representation of GenomicRange
instance generated with <cite>GenomicRange.to_dict()</cite>.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(GenomicRange) restored GenomicRange instance.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRange.merge">
<code class="sig-name descname">merge</code><span class="sig-paren">(</span><em class="sig-param">other</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRange.merge" title="Permalink to this definition">¶</a></dt>
<dd><p>Merges two genomic ranges together.</p>
<p>In case of the same genomes and chromosomes
two genomic ranges can be merged strand-nonspecifically
into one genomic range which starts where the closest to
the chromosome start genomic range starts and ends where
the closest to the chromosome end genomic range ends.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>other</strong> (<em>GenomicRange</em>) – genomic range which is merged
with self.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>(GenomicRange) new genomic range with “.” strand and</dt><dd><p>the same genome.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>InconsistentGenomesError in case of inconsistent</strong> – genomes.</p></li>
<li><p><strong>InconsistentChromosomesError in case of inconsistent</strong> – chromosomes.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRange.merge_relations">
<code class="sig-name descname">merge_relations</code><span class="sig-paren">(</span><em class="sig-param">relation</em>, <em class="sig-param">distance=0</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRange.merge_relations" title="Permalink to this definition">¶</a></dt>
<dd><p>Merges genomic ranges from relation on specified distance.</p>
<p>All genomic ranges closer to each other then the
specified distance are merged together and returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>relation</strong> (<em>str</em>) – relation type.</p></li>
<li><p><strong>distance</strong> (<em>int</em>) – distance to merge
genomic ranges in relation
(default: 0).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>(GenomicRangesList) merged genomic ranges from</dt><dd><p>relation.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><strong>ValueError in case there is no given relation</strong> – availabe for provided genomic range.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt>
<em class="property">property </em><code class="sig-name descname">name</code></dt>
<dd></dd></dl>

<dl class="method">
<dt>
<em class="property">property </em><code class="sig-name descname">sequence_file_path</code></dt>
<dd><p><cite>sequence_file_path</cite> getter.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRange.sequence_header">
<em class="property">property </em><code class="sig-name descname">sequence_header</code><a class="headerlink" href="#ortho2align.genomicranges.GenomicRange.sequence_header" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRange.to_dict">
<code class="sig-name descname">to_dict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRange.to_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns dict representation of the genomic range.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="ortho2align.genomicranges.GenomicRangesAlignment">
<em class="property">class </em><code class="sig-prename descclassname">ortho2align.genomicranges.</code><code class="sig-name descname">GenomicRangesAlignment</code><span class="sig-paren">(</span><em class="sig-param">HSPs</em>, <em class="sig-param">qrange</em>, <em class="sig-param">srange</em>, <em class="sig-param">qlen=None</em>, <em class="sig-param">slen=None</em>, <em class="sig-param">filtered_HSPs=None</em>, <em class="sig-param">relative=True</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesAlignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#ortho2align.alignment.Alignment" title="ortho2align.alignment.Alignment"><code class="xref py py-class docutils literal notranslate"><span class="pre">ortho2align.alignment.Alignment</span></code></a></p>
<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRangesAlignment.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">HSPs</em>, <em class="sig-param">qrange</em>, <em class="sig-param">srange</em>, <em class="sig-param">qlen=None</em>, <em class="sig-param">slen=None</em>, <em class="sig-param">filtered_HSPs=None</em>, <em class="sig-param">relative=True</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesAlignment.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Inits Alignment class.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul>
<li><p><strong>HSPs</strong> (<em>list</em>) – list of HSP instances;</p></li>
<li><p><strong>qrange</strong> (<em>GenomicRange</em>) – query genomic range;</p></li>
<li><p><strong>srange</strong> (<em>GenomicRange</em>) – subject genomic range;</p></li>
<li><p><strong>qlen</strong> (<em>int</em>) – query sequence length
(default: None);</p></li>
<li><p><strong>slen</strong> (<em>int</em>) – subject sequence length
(default: None);</p></li>
<li><p><strong>filtered_HSPs</strong> (<em>list</em>) –</p>
<dl class="simple">
<dt>list of HSP instances</dt><dd><p>that were filtered by some criterion</p>
</dd>
</dl>
<p>(default: None).</p>
</li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRangesAlignment.from_dict">
<em class="property">classmethod </em><code class="sig-name descname">from_dict</code><span class="sig-paren">(</span><em class="sig-param">dict_</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesAlignment.from_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Restores object from dict representation.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRangesAlignment.from_file_blast">
<em class="property">classmethod </em><code class="sig-name descname">from_file_blast</code><span class="sig-paren">(</span><em class="sig-param">file_object</em>, <em class="sig-param">qrange</em>, <em class="sig-param">srange</em>, <em class="sig-param">start_type=1</em>, <em class="sig-param">end_type='inclusive'</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesAlignment.from_file_blast" title="Permalink to this definition">¶</a></dt>
<dd><p>File parser.</p>
<p>Takes file object that corresponds to
a BLAST alignment of two sequences
produced with <cite>-outfmt 7</cite> key and
parses it into an Alignment object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_object</strong> – file object.</p></li>
<li><p><strong>qchrom</strong> (<em>str, int</em>) – query chromosome name (default: None).</p></li>
<li><p><strong>schrom</strong> (<em>str, int</em>) – subject chromosome name (default: None).</p></li>
<li><p><strong>qname</strong> (<em>str</em>) – query sequence name (default: None).</p></li>
<li><p><strong>sname</strong> (<em>str</em>) – subject sequence name (default: None).</p></li>
<li><p><strong>qstrand</strong> (<em>str</em>) – query sequence name.
One of “+”, “-“, “.” (default: “.”).</p></li>
<li><p><strong>sstrand</strong> (<em>str</em>) – subject sequence name.
One of “+”, “-“, “.” (default: “.”).</p></li>
<li><p><strong>start_type</strong> (<em>int</em>) – chromosome start coordinate.
One of 0, 1 (default: 1).</p></li>
<li><p><strong>end_type</strong> (<em>str</em>) – chromosome end inclusion type.
One of ‘inclusive’, ‘exclusive’ (default: ‘inclusive’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An Alignment instance.</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ValueError in case field 'score'</strong> – </p></li>
<li><p><strong>is not found in alignment fields.</strong> – </p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRangesAlignment.to_dict">
<code class="sig-name descname">to_dict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesAlignment.to_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Transforms instance to dict representation.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRangesAlignment.to_genomic">
<code class="sig-name descname">to_genomic</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesAlignment.to_genomic" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRangesAlignment.to_relative">
<code class="sig-name descname">to_relative</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesAlignment.to_relative" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="ortho2align.genomicranges.GenomicRangesList">
<em class="property">class </em><code class="sig-prename descclassname">ortho2align.genomicranges.</code><code class="sig-name descname">GenomicRangesList</code><span class="sig-paren">(</span><em class="sig-param">collection=[]</em>, <em class="sig-param">sequence_file_path=None</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">sortedcontainers.sortedlist.SortedKeyList</span></code></p>
<p>Represents a list of genomic ranges.</p>
<p>Genomic coordinates are 0-based exclusive:
* chromosome start coordinate is 0;
* character under genomic range coordinate</p>
<blockquote>
<div><p>is not included into the sequence.</p>
</div></blockquote>
<dl class="attribute">
<dt id="ortho2align.genomicranges.GenomicRangesList.sequence_file_path">
<code class="sig-name descname">sequence_file_path</code><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.sequence_file_path" title="Permalink to this definition">¶</a></dt>
<dd><p>path to the corresponding
genome file.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str, Path</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.genomicranges.GenomicRangesList.sequence_file">
<code class="sig-name descname">sequence_file</code><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.sequence_file" title="Permalink to this definition">¶</a></dt>
<dd><p>FastaSeqFile instance with the
corresponding genome file.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#ortho2align.genomicranges.FastaSeqFile" title="ortho2align.genomicranges.FastaSeqFile">FastaSeqFile</a></p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.genomicranges.GenomicRangesList.name_mapping">
<code class="sig-name descname">name_mapping</code><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.name_mapping" title="Permalink to this definition">¶</a></dt>
<dd><p>dictionary with keys
as genomic ranges names and values as lists of
corresponding GenomicRange instances.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dict of lists</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.genomicranges.GenomicRangesList._verbose_amount">
<code class="sig-name descname">_verbose_amount</code><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList._verbose_amount" title="Permalink to this definition">¶</a></dt>
<dd><p>how many genomic ranges to
print at maximum (positive number) (default: 10).</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="simple">
<dt>Class attributes:</dt><dd><dl class="simple">
<dt>column_names (dict of lists): column names of selected</dt><dd><p>file formats of genomic annotations.</p>
</dd>
<dt>name_columns (dict of strs): names of columns containing</dt><dd><p>information about genomic range names for selected
file formats of genomic annotations.</p>
</dd>
<dt>comments (dict of strs): comment characters for selected file</dt><dd><p>formats of genomic annotations.</p>
</dd>
<dt>name_patterns (dict of r-strs): regex patterns to extract</dt><dd><p>genomic range from the column containing the name for
selected file formats of genomic annotations.</p>
</dd>
<dt>seps (dict of strs): field separator characters for selected</dt><dd><p>file formats of genomic annotations.</p>
</dd>
<dt>fileformats (list): list of selected file formats of genomic</dt><dd><p>annotations (gtf, gff, bed3, bed6, bed12, None – user-specified).</p>
</dd>
<dt>dtypes (dict of lists): dict of lists of callables representing</dt><dd><p>dtypes of columns for selected file formats of genomic
annotations.</p>
</dd>
<dt>start_types (dict of ints): maps file format to genomic range</dt><dd><p>coordinate start type (0 or 1).</p>
</dd>
<dt>end_types (dict of strs): maps file format to genomic range</dt><dd><p>coordinate end type (‘inclusive’ or ‘exclusive’).</p>
</dd>
</dl>
</dd>
</dl>
<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRangesList.__eq__">
<code class="sig-name descname">__eq__</code><span class="sig-paren">(</span><em class="sig-param">other</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Equality magic method.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRangesList.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">collection=[]</em>, <em class="sig-param">sequence_file_path=None</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initializes GenomicRangesList instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>collection</strong> (<em>iterable</em>) – collection of GenomicRange
instances with the same genome.</p></li>
<li><p><strong>sequence_file_path</strong> (<em>str, Path</em>) – path to corresponding genome
file.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRangesList.__new__">
<em class="property">static </em><code class="sig-name descname">__new__</code><span class="sig-paren">(</span><em class="sig-param">cls</em>, <em class="sig-param">collection=[]</em>, <em class="sig-param">sequence_file_path=None</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.__new__" title="Permalink to this definition">¶</a></dt>
<dd><p>Makes new instance of GenomicRangesList.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRangesList.__repr__">
<code class="sig-name descname">__repr__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns representaion.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRangesList.__str__">
<code class="sig-name descname">__str__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns string representation.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRangesList.align_with_relation">
<code class="sig-name descname">align_with_relation</code><span class="sig-paren">(</span><em class="sig-param">relation</em>, <em class="sig-param">cores=1</em>, <em class="sig-param">verbose=False</em>, <em class="sig-param">suppress_exceptions=False</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.align_with_relation" title="Permalink to this definition">¶</a></dt>
<dd><p>Aligns each genomic range with its relation if it exists.</p>
<p>For each genomic range in self in case its’ relation
exists, aligns that genomic range with each range in
that relation using standalone blastn. Allows
multiprocessing of the alignment process.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>relation</strong> (<em>str</em>) – relation to align with.</p></li>
<li><p><strong>cores</strong> (<em>int</em>) – how many cores to use (default: 1).</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – if True, will print progress
with tqdm progress bar (default: False).</p></li>
<li><p><strong>suppress_exceptions</strong> (<em>bool</em>) – if True, will suppress
reporting exceptions to the return list (dfault: False).</p></li>
<li><p><strong>kwargs</strong> (<em>dict-like</em>) – any kwargs to pass to
GenomicRange.align representing CLI arguments
for blastn.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>the first contains GenomicRangesAlignment</dt><dd><p>instances representing alignments, the cecond contains
exceptions occured during alignment if <cite>suppress_exceptions</cite>
is False (otherwise it is empty).</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(list, list) two lists</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>EmptyGenomicRangesListError in case self genomic ranges</strong> – list is empty.</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.genomicranges.GenomicRangesList.column_names">
<code class="sig-name descname">column_names</code><em class="property"> = {'bed12': ['chrom', 'start', 'end', 'name', 'score', 'strand', 'thickstart', 'thickend', 'itemrgb', 'blockcount', 'blocksizes', 'blockstarts'], 'bed3': ['chrom', 'start', 'end'], 'bed6': ['chrom', 'start', 'end', 'name', 'score', 'strand'], 'gff': ['chrom', 'source', 'type', 'start', 'end', 'smth1', 'strand', 'smth2', 'data'], 'gtf': ['chrom', 'source', 'type', 'start', 'end', 'smth1', 'strand', 'smth2', 'data']}</em><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.column_names" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="ortho2align.genomicranges.GenomicRangesList.comments">
<code class="sig-name descname">comments</code><em class="property"> = {'bed12': '#', 'bed3': '#', 'bed6': '#', 'gff': '#', 'gtf': '#'}</em><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.comments" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="ortho2align.genomicranges.GenomicRangesList.dtypes">
<code class="sig-name descname">dtypes</code><em class="property"> = {'bed12': [&lt;class 'str'&gt;, &lt;class 'int'&gt;, &lt;class 'int'&gt;, &lt;class 'str'&gt;, &lt;class 'int'&gt;, &lt;class 'str'&gt;, &lt;class 'int'&gt;, &lt;class 'int'&gt;, &lt;class 'str'&gt;, &lt;class 'int'&gt;, &lt;function GenomicRangesList.&lt;lambda&gt;&gt;, &lt;function GenomicRangesList.&lt;lambda&gt;&gt;], 'bed3': [&lt;class 'str'&gt;, &lt;class 'int'&gt;, &lt;class 'int'&gt;], 'bed6': [&lt;class 'str'&gt;, &lt;class 'int'&gt;, &lt;class 'int'&gt;, &lt;class 'str'&gt;, &lt;class 'int'&gt;, &lt;class 'str'&gt;], 'gff': [&lt;class 'str'&gt;, &lt;class 'str'&gt;, &lt;class 'str'&gt;, &lt;class 'int'&gt;, &lt;class 'int'&gt;, &lt;class 'str'&gt;, &lt;class 'str'&gt;, &lt;class 'str'&gt;, &lt;class 'str'&gt;], 'gtf': [&lt;class 'str'&gt;, &lt;class 'str'&gt;, &lt;class 'str'&gt;, &lt;class 'int'&gt;, &lt;class 'int'&gt;, &lt;class 'str'&gt;, &lt;class 'str'&gt;, &lt;class 'str'&gt;, &lt;class 'str'&gt;]}</em><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.dtypes" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="ortho2align.genomicranges.GenomicRangesList.end_types">
<code class="sig-name descname">end_types</code><em class="property"> = {'bed12': 'exclusive', 'bed3': 'exclusive', 'bed6': 'exclusive', 'gff': 'inclusive', 'gtf': 'inclusive'}</em><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.end_types" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="attribute">
<dt id="ortho2align.genomicranges.GenomicRangesList.fileformats">
<code class="sig-name descname">fileformats</code><em class="property"> = ['gtf', 'gff', 'bed3', 'bed6', 'bed12', None]</em><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.fileformats" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRangesList.flank">
<code class="sig-name descname">flank</code><span class="sig-paren">(</span><em class="sig-param">distance=0</em>, <em class="sig-param">check_boundaries=True</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.flank" title="Permalink to this definition">¶</a></dt>
<dd><p>Flanks all genomic ranges in the list with given distance.</p>
<p>Allows correction of new start and end coordinates in case
they exceeds corresponding chromosome size.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>distance</strong> (<em>int</em>) – value to flank genomic ranges
(default: 0).</p></li>
<li><p><strong>check_boundaries</strong> (<em>bool</em>) – whether to check if
new start coordinates are not less than zero
and new end coordinates are not greater than
corresponding chromosome sizes. If True, corrects
coordinates (default: False).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>(GenomicRangesList) new genomic ranges list with flanked</dt><dd><p>and corrected genomic ranges.</p>
</dd>
</dl>
</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRangesList.from_dict">
<em class="property">classmethod </em><code class="sig-name descname">from_dict</code><span class="sig-paren">(</span><em class="sig-param">dict_</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.from_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Restores GenomicRangesList instance from dict representation.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRangesList.get_fasta">
<code class="sig-name descname">get_fasta</code><span class="sig-paren">(</span><em class="sig-param">outfileprefix</em>, <em class="sig-param">mode='split'</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.get_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Extracts sequences of genomic ranges.</p>
<p>Can extract sequences of all genomic ranges
into single file named <cite>outfileprefix</cite> + ‘.fasta’
or each sequence into separate files named
<cite>outfileprefix</cite> + genomic_range.name + ‘.fasta’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>outfileprefix</strong> (<em>str, Path</em>) – prefix of the
output file(s).</p></li>
<li><p><strong>mode</strong> (<em>str</em>) – the mode to extract sequences,
one of ‘split’, ‘bulk’ (default: ‘split’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><strong>ValueError in cases the extraction mode is not</strong> – one of ‘bulk’, ‘split’.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRangesList.get_neighbours">
<code class="sig-name descname">get_neighbours</code><span class="sig-paren">(</span><em class="sig-param">other</em>, <em class="sig-param">distance=0</em>, <em class="sig-param">relation='neighbours'</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.get_neighbours" title="Permalink to this definition">¶</a></dt>
<dd><p>Finds all neighbours of self in other at the given distance.</p>
<p>For each genomic range in self finds all ranges in other
that are closer to the range than provided distance,
forms new GenomicRangesList instance with them and assigns
it to the <cite>.relation</cite> dict under the provided relation key.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<em>GenomicRangesList</em>) – the other genomic ranges
list in which to find neighbours of self.</p></li>
<li><p><strong>distance</strong> (<em>int, float</em>) – distance at which two genomic
ranges are considered neighbours (default: 0).</p></li>
<li><p><strong>relation</strong> (<em>str</em>) – name of relation to assign all
found neighbours of each genomic range in self
(default: ‘neighbours’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRangesList.inter_ranges">
<code class="sig-name descname">inter_ranges</code><span class="sig-paren">(</span><em class="sig-param">distance=0</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.inter_ranges" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns inner genomic regions covered with no genomic ranges.</p>
<p>First merges all genomic ranges provided
in self at the given distance with
<cite>GenomicRangesList.merge</cite>, then inverses
them so as new regions do not contain any
of the previous genomic ranges. Regions
at the beginning and the end of chromosomes
are not included.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>distance</strong> – distance to merge initial
genomic ranges (default: 0).</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>new genomic ranges</dt><dd><p>list with inverted genomic ranges.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(<a class="reference internal" href="#ortho2align.genomicranges.GenomicRangesList" title="ortho2align.genomicranges.GenomicRangesList">GenomicRangesList</a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRangesList.key_func">
<em class="property">classmethod </em><code class="sig-name descname">key_func</code><span class="sig-paren">(</span><em class="sig-param">grange</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.key_func" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns key for sorting of genomic ranges.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRangesList.merge">
<code class="sig-name descname">merge</code><span class="sig-paren">(</span><em class="sig-param">distance=0</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.merge" title="Permalink to this definition">¶</a></dt>
<dd><p>Merges genomic ranges on the given distance.</p>
<p>If the distance between two genomic ranges
is smaller then the provided value, they are
merged via GenomicRange.merge and added to
the new GenomicRangesList instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>distance</strong> – distance to merge proximal
genomic ranges (default: 0).</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>(GenomicRangeList) new genomic ranges list</dt><dd><p>with all proximal genomic ranges merged.
If the initial genomic ranges list is empty,
returns empty genomic ranges list.</p>
</dd>
</dl>
</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.genomicranges.GenomicRangesList.name_columns">
<code class="sig-name descname">name_columns</code><em class="property"> = {'bed12': 'name', 'bed3': None, 'bed6': 'name', 'gff': 'data', 'gtf': 'data'}</em><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.name_columns" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt>
<em class="property">property </em><code class="sig-name descname">name_mapping</code></dt>
<dd><p>GenomicRange name mapping.</p>
<dl class="simple">
<dt>Dictionary that maps names of GenomicRange</dt><dd><p>instances to the instances themselves.</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.genomicranges.GenomicRangesList.name_patterns">
<code class="sig-name descname">name_patterns</code><em class="property"> = {'bed12': '(.*)', 'bed3': '(.*)', 'bed6': '(.*)', 'gff': 'GeneID:(\\d+)', 'gtf': 'gene_id\\s\\&quot;(\\w*)\\&quot;'}</em><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.name_patterns" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRangesList.parse_annotation">
<em class="property">classmethod </em><code class="sig-name descname">parse_annotation</code><span class="sig-paren">(</span><em class="sig-param">fileobj</em>, <em class="sig-param">fileformat=None</em>, <em class="sig-param">sequence_file_path=None</em>, <em class="sig-param">column_names=None</em>, <em class="sig-param">dtypes=None</em>, <em class="sig-param">name_column=None</em>, <em class="sig-param">name_pattern=None</em>, <em class="sig-param">comment=None</em>, <em class="sig-param">sep=None</em>, <em class="sig-param">start_type=None</em>, <em class="sig-param">end_type=None</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.parse_annotation" title="Permalink to this definition">¶</a></dt>
<dd><p>Parses annotation file into GenomicRangesList.</p>
<p>The class method for loading GenomicRangesList instance
from genomic annotation file in gtf, gff, bed3, bed6
and bed12 formats. User can provide one’s own file
formats with given parsing patterns: column_names,
dtypes, name_pattern, name_column, comment and sep. These
parsing patterns are already provided for expected file
formats. By default they are all None, so the class automatically
chooses them by provided fileformat. If fileformat is None, expects
one to provide all parsing patterns. Expects file with no header,
but possibly with commented lines.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fileobj</strong> (<em>fileobj</em>) – file-like object to parse
annotation from.</p></li>
<li><p><strong>fileformat</strong> (<em>str</em>) – type of file. Preferrably one
of ‘gtf’, ‘gff’, ‘bed3’, ‘bed6’, ‘bed12’. If
None, means custom fileformat, so column_name,
dtypes, name_pattern, comment and sep arguments must
be provided (default: None).</p></li>
<li><p><strong>sequence_file_path</strong> (<em>str, Path</em>) – path to the corresponding
genome file (default: None).</p></li>
<li><p><strong>column_names</strong> (<em>list of str</em>) – list of column names. Has
to contain ‘chrom’, ‘start’, ‘end’ values (default: None).</p></li>
<li><p><strong>dtypes</strong> (<em>list of callables</em>) – list of types of column values.
Each callable corresponds to one column (default: None).</p></li>
<li><p><strong>name_column</strong> (<em>str</em>) – name of column containing the name
of each genomic range. May be not specified so
the genomic range name will be derived from its coordinates
(defalut: None).</p></li>
<li><p><strong>name_pattern</strong> (<em>r-str</em>) – regex pattern to extract genomic range
name from column specified in name_column. Nay be not
specified so the genomic range name will be derived from
its coordinates (default: None).</p></li>
<li><p><strong>comment</strong> (<em>str</em>) – a character being used to comment lines
(default: None).</p></li>
<li><p><strong>sep</strong> (<em>str</em>) – a character used to separate fields in the
record (i.e. columns) (default: None).</p></li>
<li><p><strong>start_type</strong> (<em>int</em>) – chromosome coordinate start type: 0-based
or 1-based. One of: 0, 1 (default: None).</p></li>
<li><p><strong>end_type</strong> (<em>str</em>) – chromosome coordinate end type: inclusive or
exclusive. Inclusive means character under the end coordinate
includes in the sequence, exclusive means the opposite. One
of: ‘exclusive’, ‘inclusive’ (default: None).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>(GenomicRangesList) list of genomic ranges corresponding</dt><dd><p>to the given genome annotation.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ValueError if fileformat not one of supported</strong><strong> (</strong><strong>i.e</strong> – ‘gtf’, ‘gff’, ‘bed3’, ‘bed6’, ‘bed12’, None).</p></li>
<li><p><strong>ValueError if fileformat is None and one of annotation</strong> – patterns is None as well.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRangesList.relation_mapping">
<code class="sig-name descname">relation_mapping</code><span class="sig-paren">(</span><em class="sig-param">other</em>, <em class="sig-param">mapping</em>, <em class="sig-param">relation</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.relation_mapping" title="Permalink to this definition">¶</a></dt>
<dd><p>Maps relations of two genomic ranges lists.</p>
<p>Provided with the map of genomic ranges names in
the form of the dict with keys as names of genomic
ranges in self and values as lists of names of
genomic ranges in other, fills a relation for each
genomic range in self, that’s name is in the map keys,
with corresponding genomic ranges in other.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<em>GenomicRangesList</em>) – the other genomic ranges
list to fill relations of self genomic ranges list.</p></li>
<li><p><strong>mapping</strong> (<em>dict of lists</em>) – mapping of genomic ranges names
in self to list of genomic ranges names in other.</p></li>
<li><p><strong>relation</strong> (<em>str</em>) – the relation to fill.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.genomicranges.GenomicRangesList.seps">
<code class="sig-name descname">seps</code><em class="property"> = {'bed12': '\t', 'bed3': '\t', 'bed6': '\t', 'gff': '\t', 'gtf': '\t'}</em><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.seps" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt>
<em class="property">property </em><code class="sig-name descname">sequence_file_path</code></dt>
<dd><p>sequence_file_path getter.</p>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.genomicranges.GenomicRangesList.start_types">
<code class="sig-name descname">start_types</code><em class="property"> = {'bed12': 0, 'bed3': 0, 'bed6': 0, 'gff': 1, 'gtf': 1}</em><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.start_types" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRangesList.to_dict">
<code class="sig-name descname">to_dict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesList.to_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns dict representation.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="ortho2align.genomicranges.GenomicRangesTranscript">
<em class="property">class </em><code class="sig-prename descclassname">ortho2align.genomicranges.</code><code class="sig-name descname">GenomicRangesTranscript</code><span class="sig-paren">(</span><em class="sig-param">HSPs</em>, <em class="sig-param">alignment</em>, <em class="sig-param">score=None</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesTranscript" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#ortho2align.alignment.Transcript" title="ortho2align.alignment.Transcript"><code class="xref py py-class docutils literal notranslate"><span class="pre">ortho2align.alignment.Transcript</span></code></a></p>
<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRangesTranscript.from_dict">
<em class="property">classmethod </em><code class="sig-name descname">from_dict</code><span class="sig-paren">(</span><em class="sig-param">dict_</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesTranscript.from_dict" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.GenomicRangesTranscript.to_bed12">
<code class="sig-name descname">to_bed12</code><span class="sig-paren">(</span><em class="sig-param">mode='list'</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.GenomicRangesTranscript.to_bed12" title="Permalink to this definition">¶</a></dt>
<dd><p>Turns transcript into BED12 representation of both
query side and subject side.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mode</strong> (<em>str</em>) – how to return representation. If ‘list’ then</p></li>
<li><p><strong>list of lists, if ‘str’ then complete BED12 record.</strong></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>BED12 representation.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(list of two lists or two strs)</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="exception">
<dt id="ortho2align.genomicranges.InconsistentChromosomesError">
<em class="property">exception </em><code class="sig-prename descclassname">ortho2align.genomicranges.</code><code class="sig-name descname">InconsistentChromosomesError</code><span class="sig-paren">(</span><em class="sig-param">one_chrom</em>, <em class="sig-param">other_chrom</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.InconsistentChromosomesError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#ortho2align.genomicranges.GenomicException" title="ortho2align.genomicranges.GenomicException"><code class="xref py py-class docutils literal notranslate"><span class="pre">ortho2align.genomicranges.GenomicException</span></code></a></p>
<p>Occurs when two GenomicRange instances have different chromosomes.</p>
</dd></dl>

<dl class="exception">
<dt id="ortho2align.genomicranges.InconsistentGenomesError">
<em class="property">exception </em><code class="sig-prename descclassname">ortho2align.genomicranges.</code><code class="sig-name descname">InconsistentGenomesError</code><span class="sig-paren">(</span><em class="sig-param">one_genome</em>, <em class="sig-param">other_genome</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.InconsistentGenomesError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#ortho2align.genomicranges.GenomicException" title="ortho2align.genomicranges.GenomicException"><code class="xref py py-class docutils literal notranslate"><span class="pre">ortho2align.genomicranges.GenomicException</span></code></a></p>
<p>Occurs when two GenomicRange instances have different genomes.</p>
</dd></dl>

<dl class="exception">
<dt id="ortho2align.genomicranges.SequenceFileNotFoundError">
<em class="property">exception </em><code class="sig-prename descclassname">ortho2align.genomicranges.</code><code class="sig-name descname">SequenceFileNotFoundError</code><span class="sig-paren">(</span><em class="sig-param">instance</em>, <em class="sig-param">file_path</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.SequenceFileNotFoundError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#ortho2align.genomicranges.GenomicException" title="ortho2align.genomicranges.GenomicException"><code class="xref py py-class docutils literal notranslate"><span class="pre">ortho2align.genomicranges.GenomicException</span></code></a></p>
<p>Occurs when provided sequence file path is invalid.</p>
</dd></dl>

<dl class="class">
<dt id="ortho2align.genomicranges.SequencePath">
<em class="property">class </em><code class="sig-prename descclassname">ortho2align.genomicranges.</code><code class="sig-name descname">SequencePath</code><span class="sig-paren">(</span><em class="sig-param">path=None</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.SequencePath" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Wrapper aroudn pathlib.Path with checking of None.</p>
<dl class="attribute">
<dt id="ortho2align.genomicranges.SequencePath.path">
<code class="sig-name descname">path</code><a class="headerlink" href="#ortho2align.genomicranges.SequencePath.path" title="Permalink to this definition">¶</a></dt>
<dd><p>filesystem path.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>None or Path</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.SequencePath.__eq__">
<code class="sig-name descname">__eq__</code><span class="sig-paren">(</span><em class="sig-param">other</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.SequencePath.__eq__" title="Permalink to this definition">¶</a></dt>
<dd><p>Equality magic method.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.SequencePath.__fspath__">
<code class="sig-name descname">__fspath__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.SequencePath.__fspath__" title="Permalink to this definition">¶</a></dt>
<dd><p>Magic method to allow use with built-in <cite>open()</cite>.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.SequencePath.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">path=None</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.SequencePath.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initializes SequencePath instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>path</strong> (<em>str, bytes, Path</em>) – filesystem path
(default: None).</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.SequencePath.__repr__">
<code class="sig-name descname">__repr__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.SequencePath.__repr__" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns representation.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.SequencePath.__str__">
<code class="sig-name descname">__str__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.SequencePath.__str__" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns str representation.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.SequencePath.check_correct">
<code class="sig-name descname">check_correct</code><span class="sig-paren">(</span><em class="sig-param">foreign=None</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.SequencePath.check_correct" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks whether the path exists or it is correct.</p>
<p>Checks whether the path is not None, whether it
exists and whether it is a Path instance. If None
or doesn’t exist, reports. If not a Path instance,
casts to the Path.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>foreign</strong> (<em>object</em>) – foreign class instance
that contains SequencePath instance
(default: None).</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>FileNotSpecifiedError in case self.path is None.</strong> – </p></li>
<li><p><strong>SequenceFileNotFoundError in case self.path</strong> – points to the path that doesn’t exist.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.SequencePath.exists">
<code class="sig-name descname">exists</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.SequencePath.exists" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks whether self.path exists.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.SequencePath.open">
<code class="sig-name descname">open</code><span class="sig-paren">(</span><em class="sig-param">mode='r'</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.SequencePath.open" title="Permalink to this definition">¶</a></dt>
<dd><p>Opens corresponding file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>mode</strong> (<em>str</em>) – mode to open file
(default: ‘r’).</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(fileobj) opened file instance.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.genomicranges.SequencePath.to_dict">
<code class="sig-name descname">to_dict</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.SequencePath.to_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns dict representation.</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="ortho2align.genomicranges.align_with_relation_wrapper">
<code class="sig-prename descclassname">ortho2align.genomicranges.</code><code class="sig-name descname">align_with_relation_wrapper</code><span class="sig-paren">(</span><em class="sig-param">grange</em>, <em class="sig-param">relation</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.align_with_relation_wrapper" title="Permalink to this definition">¶</a></dt>
<dd><p>Simple wrapper around GenomicRange method for multiprocessing.</p>
</dd></dl>

<dl class="function">
<dt id="ortho2align.genomicranges.extract_taxid_mapping">
<code class="sig-prename descclassname">ortho2align.genomicranges.</code><code class="sig-name descname">extract_taxid_mapping</code><span class="sig-paren">(</span><em class="sig-param">mapping</em>, <em class="sig-param">taxid</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.extract_taxid_mapping" title="Permalink to this definition">¶</a></dt>
<dd><p>Extracts mapping of orthologs for one taxid.</p>
<p>Given the dictionary produced by <cite>ortho2align
get_orthodb_map</cite> retrieves mapping of orthologs
from query species to the provided taxid.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mapping</strong> (<em>dict</em>) – mapping of orthologs
loaded from json file produced
by <cite>ortho2align get_orthodb_map</cite>.</p></li>
<li><p><strong>taxid</strong> (<em>int, str</em>) – NCBI taxid of subject
species.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>(dict) dictionary with query species</dt><dd><p>gene ids as keys and list of subject
species orthologs’ ids as values.</p>
</dd>
</dl>
</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="ortho2align.genomicranges.fasta_reformatter">
<code class="sig-prename descclassname">ortho2align.genomicranges.</code><code class="sig-name descname">fasta_reformatter</code><span class="sig-paren">(</span><em class="sig-param">infile</em>, <em class="sig-param">outfile</em>, <em class="sig-param">total</em>, <em class="sig-param">linewidth=60</em>, <em class="sig-param">reverse=False</em>, <em class="sig-param">complement=False</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.genomicranges.fasta_reformatter" title="Permalink to this definition">¶</a></dt>
<dd><p>Reformats fasta file.</p>
<p>Allows one to reformat existing sequence in
fasta format to the specified linewidth.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>infile</strong> (<em>fileobj</em>) – input fasta file with cursor set
at the start of sequence fragment.</p></li>
<li><p><strong>outfile</strong> (<em>fileobj</em>) – output fasta file.</p></li>
<li><p><strong>total</strong> (<em>int</em>) – number of symbols to read.</p></li>
<li><p><strong>linewidth</strong> (<em>int</em>) – linewidth of new sequence
(default: 60).</p></li>
<li><p><strong>reverse</strong> (<em>bool</em>) – if True, takes sequence backwards
(default: False).</p></li>
<li><p><strong>complement</strong> (<em>bool</em>) – if True, returns complement
nucleotide sequence (default: False).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-ortho2align.orthodb">
<span id="ortho2align-orthodb-module"></span><h2>ortho2align.orthodb module<a class="headerlink" href="#module-ortho2align.orthodb" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="ortho2align.orthodb.FileOperationWrapper">
<em class="property">class </em><code class="sig-prename descclassname">ortho2align.orthodb.</code><code class="sig-name descname">FileOperationWrapper</code><span class="sig-paren">(</span><em class="sig-param">prefix</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.orthodb.FileOperationWrapper" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Provides context manager and dictionary-
based file handler for split_table function.</p>
<dl class="attribute">
<dt id="ortho2align.orthodb.FileOperationWrapper.file_dict">
<code class="sig-name descname">file_dict</code><a class="headerlink" href="#ortho2align.orthodb.FileOperationWrapper.file_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>handler of opened files.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.orthodb.FileOperationWrapper.prefix">
<code class="sig-name descname">prefix</code><a class="headerlink" href="#ortho2align.orthodb.FileOperationWrapper.prefix" title="Permalink to this definition">¶</a></dt>
<dd><p>file prefix for each of
the opened files.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str, Path</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.orthodb.FileOperationWrapper.__getitem__">
<code class="sig-name descname">__getitem__</code><span class="sig-paren">(</span><em class="sig-param">key</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.orthodb.FileOperationWrapper.__getitem__" title="Permalink to this definition">¶</a></dt>
<dd><p>Item getter.</p>
<p>In case file associated with the key
doesn’t exist in <cite>self.file_dict</cite>, it
is opened for writing and added to
<cite>self.file_dict</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>key</strong> (<em>str</em>) – a key to file instance.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p>file instance opened at the location
with following scheme:</p>
<blockquote>
<div><p>self.prefix + “_” + key</p>
</div></blockquote>
</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.orthodb.FileOperationWrapper.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">prefix</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.orthodb.FileOperationWrapper.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initializes FileOperationWrapper object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>prefix</strong> (<em>str, Path</em>) – file prefix.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.orthodb.FileOperationWrapper.__setitem__">
<code class="sig-name descname">__setitem__</code><span class="sig-paren">(</span><em class="sig-param">key</em>, <em class="sig-param">item</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.orthodb.FileOperationWrapper.__setitem__" title="Permalink to this definition">¶</a></dt>
<dd><p>Item setter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> (<em>str</em>) – a key to file instance.</p></li>
<li><p><strong>item</strong> (<em>file-like</em>) – file instance.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="ortho2align.orthodb.filter_odb_file">
<code class="sig-prename descclassname">ortho2align.orthodb.</code><code class="sig-name descname">filter_odb_file</code><span class="sig-paren">(</span><em class="sig-param">file_suffix</em>, <em class="sig-param">odb_path</em>, <em class="sig-param">odb_prefix</em>, <em class="sig-param">column</em>, <em class="sig-param">terms</em>, <em class="sig-param">column_list</em>, <em class="sig-param">cache_path</em>, <em class="sig-param">code</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.orthodb.filter_odb_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Filters OrthoDB table by terms in given column.</p>
<p>Leaves only those lines that contain one of
the terms in specified column.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_suffix</strong> (<em>str</em>) – suffix of OrthoDB table to be filtered.</p></li>
<li><p><strong>odb_path</strong> (<em>str, Path</em>) – path to OrthoDB folder.</p></li>
<li><p><strong>odb_prefix</strong> (<em>str</em>) – prefix of OrthoDB files.</p></li>
<li><p><strong>column</strong> (<em>str</em>) – column name to filter on.</p></li>
<li><p><strong>terms</strong> (<em>set</em>) – a set of terms to find in column.</p></li>
<li><p><strong>column_list</strong> (<em>list</em>) – list of column names.</p></li>
<li><p><strong>cache_path</strong> (<em>str, Path</em>) – a path to cache directory.</p></li>
<li><p><strong>code</strong> (<em>str</em>) – a code to write at the end of the filtered
file.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>cache_path + “/” + odb_prefix + odb_suffix +
“_” + column + “-” + code</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(Path) a path to queried file in the following scheme</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="ortho2align.orthodb.filter_table">
<code class="sig-prename descclassname">ortho2align.orthodb.</code><code class="sig-name descname">filter_table</code><span class="sig-paren">(</span><em class="sig-param">infile</em>, <em class="sig-param">outfile</em>, <em class="sig-param">column</em>, <em class="sig-param">terms</em>, <em class="sig-param">column_list</em>, <em class="sig-param">sep</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.orthodb.filter_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Filters table by terms in given column.</p>
<p>Leaves only those lines that contain one of
the terms in specified column.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>infile</strong> (<em>file-like</em>) – input file.</p></li>
<li><p><strong>outfile</strong> (<em>file-like</em>) – output file.</p></li>
<li><p><strong>column</strong> (<em>str</em>) – column name to query on.</p></li>
<li><p><strong>terms</strong> (<em>set</em>) – a set of terms to find in column.</p></li>
<li><p><strong>column_list</strong> (<em>list</em>) – list of column names.</p></li>
<li><p><strong>sep</strong> (<em>str</em>) – field separator.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="ortho2align.orthodb.load_odb_file">
<code class="sig-prename descclassname">ortho2align.orthodb.</code><code class="sig-name descname">load_odb_file</code><span class="sig-paren">(</span><em class="sig-param">file_suffix</em>, <em class="sig-param">odb_path</em>, <em class="sig-param">odb_prefix</em>, <em class="sig-param">colnames</em>, <em class="sig-param">dtype</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.orthodb.load_odb_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads OrthoDB table into pandas.DataFrame.</p>
<p>The loaded file is <cite>odb_path + odb_prefix + file_suffix</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_suffix</strong> (<em>str</em>) – suffix of OrthoDB table to be loaded.</p></li>
<li><p><strong>odb_path</strong> (<em>str, Path</em>) – path to OrthoDB folder.</p></li>
<li><p><strong>odb_prefix</strong> (<em>str</em>) – prefix of OrthoDB files.</p></li>
<li><p><strong>colnames</strong> (<em>list</em>) – list of column names of OrthoDB table.</p></li>
<li><p><strong>dtype</strong> (<em>dict</em>) – dictionary of column dtypes as {‘colname’: ‘coldtype’}.</p></li>
<li><p><strong>**kwargs</strong> – any other arguments to be passed to <cite>pandas.read_csv()</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>pandas.DataFrame instance with loaded OrthoDB table.
Columns are named as in <cite>colnames</cite>, column dtypes are
as in <cite>dtypes</cite>.</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><strong>ValueError in case file doesn't exist.</strong> – </p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="ortho2align.orthodb.load_table">
<code class="sig-prename descclassname">ortho2align.orthodb.</code><code class="sig-name descname">load_table</code><span class="sig-paren">(</span><em class="sig-param">path</em>, <em class="sig-param">colnames</em>, <em class="sig-param">dtype</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.orthodb.load_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads tab-separated table into memory.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> (<em>str, Path</em>) – path to table.</p></li>
<li><p><strong>colnames</strong> (<em>list</em>) – list of column names of OrthoDB table.</p></li>
<li><p><strong>dtype</strong> (<em>dict</em>) – dictionary of column dtypes as {‘colname’: ‘coldtype’}.</p></li>
<li><p><strong>**kwargs</strong> – any other arguments to be passed to <cite>pandas.read_csv()</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>pandas.DataFrame instance with loaded table.
Columns are named as in <cite>colnames</cite>, column dtypes are
as in <cite>dtypes</cite>.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="ortho2align.orthodb.query_cached_odb_file">
<code class="sig-prename descclassname">ortho2align.orthodb.</code><code class="sig-name descname">query_cached_odb_file</code><span class="sig-paren">(</span><em class="sig-param">file_suffix</em>, <em class="sig-param">odb_path</em>, <em class="sig-param">odb_prefix</em>, <em class="sig-param">column</em>, <em class="sig-param">term</em>, <em class="sig-param">column_list</em>, <em class="sig-param">cache_path</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.orthodb.query_cached_odb_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Queries cached OrthoDB file.</p>
<p>In case of the first query of the given
file in given column on given term
it is queried and written to cache
directory.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_suffix</strong> (<em>str</em>) – suffix of OrthoDB table to be loaded.</p></li>
<li><p><strong>odb_path</strong> (<em>str, Path</em>) – path to OrthoDB folder.</p></li>
<li><p><strong>odb_prefix</strong> (<em>str</em>) – prefix of OrthoDB files.</p></li>
<li><p><strong>column</strong> (<em>str</em>) – column name to query on.</p></li>
<li><p><strong>term</strong> (<em>str</em>) – a term to find in column.</p></li>
<li><p><strong>column_list</strong> (<em>list</em>) – list of column names.</p></li>
<li><p><strong>cache_path</strong> (<em>str, Path</em>) – a path to cache directory.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>cache_path + “/” + odb_prefix + odb_suffix +
“_” + column + “_” + term</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(Path) a path to queried file in the following scheme</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="ortho2align.orthodb.query_table">
<code class="sig-prename descclassname">ortho2align.orthodb.</code><code class="sig-name descname">query_table</code><span class="sig-paren">(</span><em class="sig-param">infile</em>, <em class="sig-param">outfile</em>, <em class="sig-param">column</em>, <em class="sig-param">term</em>, <em class="sig-param">column_list</em>, <em class="sig-param">sep</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.orthodb.query_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Queries table column by term.</p>
<p>Finds all lines in infile, that contain
<cite>term</cite> in column named <cite>column</cite> and writes
that lines to outfile with excluded column
that was queried on.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>infile</strong> (<em>file-like</em>) – input file.</p></li>
<li><p><strong>outfile</strong> (<em>file-like</em>) – output file.</p></li>
<li><p><strong>column</strong> (<em>str</em>) – column name to query on.</p></li>
<li><p><strong>term</strong> (<em>str</em>) – a term to find in column.</p></li>
<li><p><strong>column_list</strong> (<em>list</em>) – list of column names.</p></li>
<li><p><strong>sep</strong> (<em>str</em>) – field separator.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="ortho2align.orthodb.split_odb_file">
<code class="sig-prename descclassname">ortho2align.orthodb.</code><code class="sig-name descname">split_odb_file</code><span class="sig-paren">(</span><em class="sig-param">file_suffix</em>, <em class="sig-param">odb_path</em>, <em class="sig-param">odb_prefix</em>, <em class="sig-param">column</em>, <em class="sig-param">column_list</em>, <em class="sig-param">cache_path</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.orthodb.split_odb_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Splits OrthoDB table on many files based on terms in specific column.</p>
<p>For each unique term in column a file is created
and lines containing specific term in the column
are written to corresponding files. The column that
was filtered on is omitted. New files are named
in the following scheme:
cache_path “/” + odb_prefix + file_suffix + “_” +
column + “_” + term.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_suffix</strong> (<em>str</em>) – suffix of OrthoDB table to be splitted.</p></li>
<li><p><strong>odb_path</strong> (<em>str, Path</em>) – path to OrthoDB folder.</p></li>
<li><p><strong>odb_prefix</strong> (<em>str</em>) – prefix of OrthoDB files.</p></li>
<li><p><strong>column</strong> (<em>str</em>) – column name to split on.</p></li>
<li><p><strong>column_list</strong> (<em>list</em>) – list of column names.</p></li>
<li><p><strong>cache_path</strong> (<em>str, Path</em>) – a path to cache directory.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="ortho2align.orthodb.split_table">
<code class="sig-prename descclassname">ortho2align.orthodb.</code><code class="sig-name descname">split_table</code><span class="sig-paren">(</span><em class="sig-param">infile</em>, <em class="sig-param">outprefix</em>, <em class="sig-param">column</em>, <em class="sig-param">column_list</em>, <em class="sig-param">sep</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.orthodb.split_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Splits table on many files based on terms in specific column.</p>
<p>For each unique term in column a file is created
and lines containing specific term in the column
are written to corresponding files. The column that
was filtered on is omitted. New files are named
in the following scheme:
outprefix + “_” + term.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>infile</strong> (<em>file-like</em>) – input file.</p></li>
<li><p><strong>outprefix</strong> (<em>str, Path</em>) – output files prefix.</p></li>
<li><p><strong>column</strong> (<em>str</em>) – column name to filter on.</p></li>
<li><p><strong>column_list</strong> (<em>list</em>) – list of column names.</p></li>
<li><p><strong>sep</strong> (<em>str</em>) – field separator.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None.</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-ortho2align.parallel">
<span id="ortho2align-parallel-module"></span><h2>ortho2align.parallel module<a class="headerlink" href="#module-ortho2align.parallel" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="ortho2align.parallel.NonExceptionalProcessPool">
<em class="property">class </em><code class="sig-prename descclassname">ortho2align.parallel.</code><code class="sig-name descname">NonExceptionalProcessPool</code><span class="sig-paren">(</span><em class="sig-param">max_workers=None</em>, <em class="sig-param">verbose=True</em>, <em class="sig-param">suppress_exceptions=False</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.parallel.NonExceptionalProcessPool" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Provides multiprocessing with no exceptions invoked.</p>
<p>Based on concurrent.futures.ProcessPoolExecutor
mimics multiprocessing.Pool API for mapping
function to an iterable containing arguments.</p>
<dl class="attribute">
<dt id="ortho2align.parallel.NonExceptionalProcessPool.executor">
<code class="sig-name descname">executor</code><a class="headerlink" href="#ortho2align.parallel.NonExceptionalProcessPool.executor" title="Permalink to this definition">¶</a></dt>
<dd><p>multiprocessing instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>ProcessPoolExecutor</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.parallel.NonExceptionalProcessPool.max_workers">
<code class="sig-name descname">max_workers</code><a class="headerlink" href="#ortho2align.parallel.NonExceptionalProcessPool.max_workers" title="Permalink to this definition">¶</a></dt>
<dd><p>number of max_workers for executor
(default: None, i.e. os.cpu_count()).</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.parallel.NonExceptionalProcessPool.verbose">
<code class="sig-name descname">verbose</code><a class="headerlink" href="#ortho2align.parallel.NonExceptionalProcessPool.verbose" title="Permalink to this definition">¶</a></dt>
<dd><p>if True will show progress via tqdm
(default: True).</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="ortho2align.parallel.NonExceptionalProcessPool.suppress_exceptions">
<code class="sig-name descname">suppress_exceptions</code><a class="headerlink" href="#ortho2align.parallel.NonExceptionalProcessPool.suppress_exceptions" title="Permalink to this definition">¶</a></dt>
<dd><p>if True, exceptions won’t be
recordered (default: False).</p>
</dd></dl>

<p class="rubric">Notes</p>
<p>There are 4 methods for multiprocessing: <cite>map</cite>, <cite>starmap</cite>,
<cite>map_async</cite>, <cite>starmap_async</cite>. <cite>map</cite> and <cite>map_async</cite> allows
multiprocessing of one-argument function execution.
<cite>starmap</cite> and <cite>starmap_async</cite> allows multiprocessing of
many-arguments function execution. Each method takes
function and iterable containing arguments one argument
(or group of arguments) per item. Iterable can be a collection,
generator and coroutine as well. <cite>map</cite> and <cite>starmap</cite>
return results and exceptions in the same order as arguments
were passed by an iterable. Async versions of these functions
do not guarantee that.</p>
<dl>
<dt>Usage:</dt><dd><dl>
<dt>def foo(x):</dt><dd><dl class="simple">
<dt>for _ in range(2000000):</dt><dd><p>continue</p>
</dd>
<dt>if x % 10 == 0:</dt><dd><p>raise ValueError(f”{x} is divided by ten.”)</p>
</dd>
</dl>
<p>return x</p>
</dd>
<dt>with NonExceptionalProcessPool(max_workers=5) as p:</dt><dd><p>results, exceptions = p.map(foo, range(100))</p>
</dd>
</dl>
</dd>
<dt>Explanation of usage:</dt><dd><p>All results returned by the functions are stored in
the results list. All the exceptions are stored in
the exceptions list unless <cite>suppress_exceptions</cite> is
True.</p>
</dd>
</dl>
<dl class="method">
<dt id="ortho2align.parallel.NonExceptionalProcessPool.__enter__">
<code class="sig-name descname">__enter__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.parallel.NonExceptionalProcessPool.__enter__" title="Permalink to this definition">¶</a></dt>
<dd><p>Context manager enter.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.parallel.NonExceptionalProcessPool.__exit__">
<code class="sig-name descname">__exit__</code><span class="sig-paren">(</span><em class="sig-param">exc_type</em>, <em class="sig-param">exc_val</em>, <em class="sig-param">exc_tb</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.parallel.NonExceptionalProcessPool.__exit__" title="Permalink to this definition">¶</a></dt>
<dd><p>Context manager exit.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.parallel.NonExceptionalProcessPool.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">max_workers=None</em>, <em class="sig-param">verbose=True</em>, <em class="sig-param">suppress_exceptions=False</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.parallel.NonExceptionalProcessPool.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initializes NonExceptionalProcessPool instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>max_workers</strong> (<em>int</em>) – number of max_workers for executor
(default: None, i.e. os.cpu_count()).</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – if True will show progress via tqdm
(default: True).</p></li>
<li><p><strong>suppress_exceptions</strong> – if True, exceptions won’t be
recordered (default: False).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.parallel.NonExceptionalProcessPool._map_async">
<code class="sig-name descname">_map_async</code><span class="sig-paren">(</span><em class="sig-param">func</em>, <em class="sig-param">iterable</em>, <em class="sig-param">asynchronous=True</em>, <em class="sig-param">applier=&lt;function apply&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.parallel.NonExceptionalProcessPool._map_async" title="Permalink to this definition">¶</a></dt>
<dd><p>Helper function to sustain API.</p>
<p>If <a href="#id1"><span class="problematic" id="id2">`</span></a>self.verbose`is True, than the progress will be
reported with tqdm.tqdm progress bar.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>func</strong> (<em>function</em>) – function object to multiprocess.
Can’t accept lambdas.</p></li>
<li><p><strong>iterable</strong> (<em>iterable</em>) – collection, range, map, zip,
generator, coroutine object containing arguments
to pass to func. The iterable is guaranteed to
be not transformed to list so it is safe to pass
generators and similar objects.</p></li>
<li><p><strong>asynchronous</strong> (<em>bool</em>) – whether to process calls
asynchronously or not (default: True).</p></li>
<li><p><strong>applier</strong> (<em>function</em>) – how to apply arguments in
iterable to func. One of <cite>apply</cite>, <cite>starapply</cite>
(default: <cite>apply</cite>).</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(list, list) two lists. The first contains results
of successful call, the second contains exceptions
if <cite>self.suppress_exceptions</cite> is False.</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><strong>ValueError in case provided iterable is not iterable.</strong> – </p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.parallel.NonExceptionalProcessPool.map">
<code class="sig-name descname">map</code><span class="sig-paren">(</span><em class="sig-param">func</em>, <em class="sig-param">iterable</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.parallel.NonExceptionalProcessPool.map" title="Permalink to this definition">¶</a></dt>
<dd><p>Maps one-argument function to an iterable.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>func</strong> (<em>function</em>) – function object to multiprocess.
Can’t accept lambdas.</p></li>
<li><p><strong>iterable</strong> (<em>iterable</em>) – collection, range, map, zip,
generator, coroutine object containing arguments
to pass to func. The iterable is guaranteed to
be not transformed to list so it is safe to pass
generators and similar objects.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(list, list) two lists. The first contains results
of successful call, the second contains exceptions
if <cite>self.suppress_exceptions</cite> is False. Results
and exceptions are returned in the same order as
corresponding arguments were passed via iterable.</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><strong>ValueError in case provided iterable is not iterable.</strong> – </p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.parallel.NonExceptionalProcessPool.map_async">
<code class="sig-name descname">map_async</code><span class="sig-paren">(</span><em class="sig-param">func</em>, <em class="sig-param">iterable</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.parallel.NonExceptionalProcessPool.map_async" title="Permalink to this definition">¶</a></dt>
<dd><p>Maps one-argument function to an iterable in asynchronous way.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>func</strong> (<em>function</em>) – function object to multiprocess.
Can’t accept lambdas.</p></li>
<li><p><strong>iterable</strong> (<em>iterable</em>) – collection, range, map, zip,
generator, coroutine object containing arguments
to pass to func. The iterable is guaranteed to
be not transformed to list so it is safe to pass
generators and similar objects.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(list, list) two lists. The first contains results
of successful call, the second contains exceptions
if <cite>self.suppress_exceptions</cite> is False. Results
and exceptions are not guaranteed to be returned
in the same order as corresponding arguments were
passed via iterable.</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><strong>ValueError in case provided iterable is not iterable.</strong> – </p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.parallel.NonExceptionalProcessPool.shutdown">
<code class="sig-name descname">shutdown</code><span class="sig-paren">(</span><em class="sig-param">wait=True</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.parallel.NonExceptionalProcessPool.shutdown" title="Permalink to this definition">¶</a></dt>
<dd><p>Explicitly shuts down self.executor.</p>
</dd></dl>

<dl class="method">
<dt id="ortho2align.parallel.NonExceptionalProcessPool.starmap">
<code class="sig-name descname">starmap</code><span class="sig-paren">(</span><em class="sig-param">func</em>, <em class="sig-param">iterable</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.parallel.NonExceptionalProcessPool.starmap" title="Permalink to this definition">¶</a></dt>
<dd><p>Maps many-argument function to an iterable.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>func</strong> (<em>function</em>) – function object to multiprocess.
Can’t accept lambdas.</p></li>
<li><p><strong>iterable</strong> (<em>iterable</em>) – collection, range, map, zip,
generator, coroutine object containing arguments
to pass to func. The iterable is guaranteed to
be not transformed to list so it is safe to pass
generators and similar objects.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(list, list) two lists. The first contains results
of successful call, the second contains exceptions
if <cite>self.suppress_exceptions</cite> is False. Results
and exceptions are returned in the same order as
corresponding arguments were passed via iterable.</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><strong>ValueError in case provided iterable is not iterable.</strong> – </p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="ortho2align.parallel.NonExceptionalProcessPool.starmap_async">
<code class="sig-name descname">starmap_async</code><span class="sig-paren">(</span><em class="sig-param">func</em>, <em class="sig-param">iterable</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.parallel.NonExceptionalProcessPool.starmap_async" title="Permalink to this definition">¶</a></dt>
<dd><p>Maps multi-argument function to an iterable in asynchronous way.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>func</strong> (<em>function</em>) – function object to multiprocess.
Can’t accept lambdas.</p></li>
<li><p><strong>iterable</strong> (<em>iterable</em>) – collection, range, map, zip,
generator, coroutine object containing arguments
to pass to func. The iterable is guaranteed to
be not transformed to list so it is safe to pass
generators and similar objects.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>(list, list) two lists. The first contains results
of successful call, the second contains exceptions
if <cite>self.suppress_exceptions</cite> is False. Results
and exceptions are not guaranteed to be returned
in the same order as corresponding arguments were
passed via iterable.</p>
</dd>
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><strong>ValueError in case provided iterable is not iterable.</strong> – </p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="ortho2align.parallel.apply">
<code class="sig-prename descclassname">ortho2align.parallel.</code><code class="sig-name descname">apply</code><span class="sig-paren">(</span><em class="sig-param">func</em>, <em class="sig-param">arg</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.parallel.apply" title="Permalink to this definition">¶</a></dt>
<dd><p>Applies one arg to the func.</p>
</dd></dl>

<dl class="function">
<dt id="ortho2align.parallel.starapply">
<code class="sig-prename descclassname">ortho2align.parallel.</code><code class="sig-name descname">starapply</code><span class="sig-paren">(</span><em class="sig-param">func</em>, <em class="sig-param">args</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.parallel.starapply" title="Permalink to this definition">¶</a></dt>
<dd><p>Applies starred args to the func.</p>
</dd></dl>

<dl class="function">
<dt id="ortho2align.parallel.starstarapply">
<code class="sig-prename descclassname">ortho2align.parallel.</code><code class="sig-name descname">starstarapply</code><span class="sig-paren">(</span><em class="sig-param">func</em>, <em class="sig-param">arg_container</em><span class="sig-paren">)</span><a class="headerlink" href="#ortho2align.parallel.starstarapply" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</div>
<div class="section" id="module-ortho2align">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-ortho2align" title="Permalink to this headline">¶</a></h2>
</div>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Dmitry Mylarshchikov

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>